{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup\n",
    "This IPython notebook will walk through the steps of characterizing iModulons through the semi-automated tools in PyModulon. You will need:\n",
    "\n",
    "* M and A matrices\n",
    "* Expression data (e.g. `log_tpm_norm.csv`)\n",
    "* Gene table and KEGG/GO annotations (Generated in `gene_annotation.ipynb`)\n",
    "* Sample table, with a column for `project` and `condition`\n",
    "* TRN file\n",
    "\n",
    "Optional:\n",
    "* iModulon table (if you already have some characterized iModulons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.core import IcaData\n",
    "from pymodulon.plotting import *\n",
    "from os import path\n",
    "import pandas as pd\n",
    "import re\n",
    "from Bio.KEGG import REST\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enter the location of your data here\n",
    "data_dir = path.join('../..','data','processed', 'modulome')\n",
    "\n",
    "# GO and KEGG annotations are in the 'external' folder\n",
    "external_data = path.join('../..','data','external', 'modulome')\n",
    "interim_data = path.join('../..','data','interim', 'modulome')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check your sample table (i.e. metadata file)\n",
    "Your metadata file will probably have a lot of columns, most of which you may not care about. Feel free to save a secondary copy of your metadata file with only columns that seem relevant to you. The two most important columns are:\n",
    "1. `project`\n",
    "2. `condition`\n",
    "\n",
    "Make sure that these columns exist in your metadata file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>experiment</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Control1-MSG</th>\n",
       "      <td>Control</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Control2-MSG</th>\n",
       "      <td>Control</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27-J1074-RNA-Cryo-MM</th>\n",
       "      <td>Batchcontrol</td>\n",
       "      <td>Batchcontrol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28-J1074-RNA-Cryo-MM</th>\n",
       "      <td>Batchcontrol</td>\n",
       "      <td>Batchcontrol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7_Plate_A_MS</th>\n",
       "      <td>Solid culture</td>\n",
       "      <td>Plate_A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            project     condition\n",
       "experiment                                       \n",
       "Control1-MSG                Control       control\n",
       "Control2-MSG                Control       control\n",
       "27-J1074-RNA-Cryo-MM   Batchcontrol  Batchcontrol\n",
       "28-J1074-RNA-Cryo-MM   Batchcontrol  Batchcontrol\n",
       "7_Plate_A_MS          Solid culture       Plate_A"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata = pd.read_csv(path.join(data_dir,'metadata_sorted.tsv'),index_col=0,sep='\\t')\n",
    "df_metadata[['project','condition']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "print(df_metadata.project.notnull().all())\n",
    "print(df_metadata.condition.notnull().all())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Organize the data\n",
    "We need to modify the sample order of the A.csv file because it governs the order of samples in the dashboard. We will use the `df_metadata` to do this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read A.csv\n",
    "A = pd.read_csv(path.join(data_dir,'A.csv'),index_col=0)\n",
    "A = A.reindex(columns=df_metadata['sample_id'])\n",
    "A.to_csv(path.join(data_dir,'A_ordered.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read log_tpm_norm.csv\n",
    "log_tpm = pd.read_csv(path.join(data_dir,'log_tpm_norm.csv'),index_col=0)\n",
    "log_tpm = log_tpm.reindex(columns=df_metadata['sample_id'])\n",
    "log_tpm.to_csv(path.join(data_dir,'log_tpm_norm_ordered.csv'))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the data\n",
    "You're now ready to load your IcaData object!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data = IcaData(M = path.join(data_dir,'M.csv'),\n",
    "                   A = path.join(data_dir,'A_ordered.csv'),\n",
    "                   X = path.join(data_dir,'log_tpm_norm_ordered.csv'),\n",
    "                   gene_table = path.join(data_dir,'gene_info6.csv'),\n",
    "                   sample_table = path.join(data_dir,'metadata_sorted.tsv'),\n",
    "                   threshold_method='kmeans')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functional iModulons"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GO annotations and KEGG pathways/modules were generated in the 1_create_the_gene_table.ipynb notebook. Enrichments will be calculated in this notebook, and further curated in the 3_manual_iModulon_curation notebook."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GO Enrichments"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First load the Gene Ontology annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>GOs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>XNR_RS00270</td>\n",
       "      <td>mannosyltransferase activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>XNR_RS00270</td>\n",
       "      <td>molecular_function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>XNR_RS00270</td>\n",
       "      <td>catalytic activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>XNR_RS00270</td>\n",
       "      <td>dolichyl-phosphate beta-D-mannosyltransferase ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>XNR_RS00270</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      locus_tag                                                GOs\n",
       "54  XNR_RS00270                       mannosyltransferase activity\n",
       "55  XNR_RS00270                                 molecular_function\n",
       "56  XNR_RS00270                                 catalytic activity\n",
       "57  XNR_RS00270  dolichyl-phosphate beta-D-mannosyltransferase ...\n",
       "58  XNR_RS00270                                 cellular_component"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_GO = pd.read_csv(path.join(external_data,'GO_annotations_curated.csv'),index_col=0)\n",
    "DF_GO.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change 'locus_tag' to 'gene_id'\n",
    "DF_GO = DF_GO.rename(columns={'locus_tag':'gene_id'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO_enrich = ica_data.compute_annotation_enrichment(DF_GO,'GOs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>GOs</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>N-alpha,N-alpha,N-alpha-trimethyl-L-histidine ...</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.016441</td>\n",
       "      <td>0.037736</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.072727</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>amino-acid betaine biosynthetic process</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.016441</td>\n",
       "      <td>0.037736</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.072727</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>amino-acid betaine metabolic process</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.016441</td>\n",
       "      <td>0.037736</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.072727</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>ammonium ion metabolic process</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.016441</td>\n",
       "      <td>0.037736</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.072727</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>cellular modified histidine biosynthetic process</td>\n",
       "      <td>0.000080</td>\n",
       "      <td>0.016441</td>\n",
       "      <td>0.037736</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.072727</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>71</td>\n",
       "      <td>ncRNA processing</td>\n",
       "      <td>0.002176</td>\n",
       "      <td>0.051520</td>\n",
       "      <td>0.036364</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>4.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>110.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>71</td>\n",
       "      <td>rRNA processing</td>\n",
       "      <td>0.002453</td>\n",
       "      <td>0.057121</td>\n",
       "      <td>0.027273</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.048000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>110.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>71</td>\n",
       "      <td>RNA processing</td>\n",
       "      <td>0.002463</td>\n",
       "      <td>0.057121</td>\n",
       "      <td>0.036364</td>\n",
       "      <td>0.129032</td>\n",
       "      <td>0.056738</td>\n",
       "      <td>4.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>110.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>71</td>\n",
       "      <td>RNA metabolic process</td>\n",
       "      <td>0.002926</td>\n",
       "      <td>0.067171</td>\n",
       "      <td>0.045455</td>\n",
       "      <td>0.094340</td>\n",
       "      <td>0.061350</td>\n",
       "      <td>5.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>110.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>247</th>\n",
       "      <td>71</td>\n",
       "      <td>rRNA metabolic process</td>\n",
       "      <td>0.002979</td>\n",
       "      <td>0.067705</td>\n",
       "      <td>0.027273</td>\n",
       "      <td>0.187500</td>\n",
       "      <td>0.047619</td>\n",
       "      <td>3.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>110.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>248 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     imodulon                                                GOs    pvalue  \\\n",
       "0           0  N-alpha,N-alpha,N-alpha-trimethyl-L-histidine ...  0.000080   \n",
       "1           0            amino-acid betaine biosynthetic process  0.000080   \n",
       "2           0               amino-acid betaine metabolic process  0.000080   \n",
       "3           0                     ammonium ion metabolic process  0.000080   \n",
       "4           0   cellular modified histidine biosynthetic process  0.000080   \n",
       "..        ...                                                ...       ...   \n",
       "243        71                                   ncRNA processing  0.002176   \n",
       "244        71                                    rRNA processing  0.002453   \n",
       "245        71                                     RNA processing  0.002463   \n",
       "246        71                              RNA metabolic process  0.002926   \n",
       "247        71                             rRNA metabolic process  0.002979   \n",
       "\n",
       "       qvalue  precision    recall   f1score   TP  target_set_size  \\\n",
       "0    0.016441   0.037736  1.000000  0.072727  2.0              2.0   \n",
       "1    0.016441   0.037736  1.000000  0.072727  2.0              2.0   \n",
       "2    0.016441   0.037736  1.000000  0.072727  2.0              2.0   \n",
       "3    0.016441   0.037736  1.000000  0.072727  2.0              2.0   \n",
       "4    0.016441   0.037736  1.000000  0.072727  2.0              2.0   \n",
       "..        ...        ...       ...       ...  ...              ...   \n",
       "243  0.051520   0.036364  0.133333  0.057143  4.0             30.0   \n",
       "244  0.057121   0.027273  0.200000  0.048000  3.0             15.0   \n",
       "245  0.057121   0.036364  0.129032  0.056738  4.0             31.0   \n",
       "246  0.067171   0.045455  0.094340  0.061350  5.0             53.0   \n",
       "247  0.067705   0.027273  0.187500  0.047619  3.0             16.0   \n",
       "\n",
       "     imodulon_size  \n",
       "0             53.0  \n",
       "1             53.0  \n",
       "2             53.0  \n",
       "3             53.0  \n",
       "4             53.0  \n",
       "..             ...  \n",
       "243          110.0  \n",
       "244          110.0  \n",
       "245          110.0  \n",
       "246          110.0  \n",
       "247          110.0  \n",
       "\n",
       "[248 rows x 10 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_GO_enrich"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## KEGG Enrichments"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load KEGG mapping\n",
    "The `kegg_mapping.csv` file contains KEGG orthologies, pathways, modules, and reactions. Only pathways and modules are relevant to iModulon characterization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['KEGG_ko' 'KEGG_Pathway' 'KEGG_Module' 'KEGG_Reaction']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>database</th>\n",
       "      <th>kegg_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XNR_RS30570</td>\n",
       "      <td>KEGG_ko</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XNR_RS00010</td>\n",
       "      <td>KEGG_ko</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XNR_RS00015</td>\n",
       "      <td>KEGG_ko</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XNR_RS00020</td>\n",
       "      <td>KEGG_ko</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XNR_RS00025</td>\n",
       "      <td>KEGG_ko</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_id database kegg_id\n",
       "0  XNR_RS30570  KEGG_ko       -\n",
       "1  XNR_RS00010  KEGG_ko       -\n",
       "2  XNR_RS00015  KEGG_ko       -\n",
       "3  XNR_RS00020  KEGG_ko       -\n",
       "4  XNR_RS00025  KEGG_ko       -"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_KEGG = pd.read_csv(path.join(external_data,'kegg_mapping.csv'),index_col=0)\n",
    "print(DF_KEGG.database.unique())\n",
    "DF_KEGG.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg_pathways = DF_KEGG[DF_KEGG.database == 'KEGG_Pathway']\n",
    "kegg_modules = DF_KEGG[DF_KEGG.database == 'KEGG_Module']"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform enrichment\n",
    "Uses the `compute_annotation_enrichment` function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_pathway_enrich = ica_data.compute_annotation_enrichment(kegg_pathways,'kegg_id')\n",
    "DF_module_enrich = ica_data.compute_annotation_enrichment(kegg_modules,'kegg_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>map00020</td>\n",
       "      <td>0.000345</td>\n",
       "      <td>0.083798</td>\n",
       "      <td>0.075472</td>\n",
       "      <td>0.105263</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>4.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>map00525</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005310</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.173913</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>map01053</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.005310</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>map01130</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.005310</td>\n",
       "      <td>0.350000</td>\n",
       "      <td>0.021875</td>\n",
       "      <td>0.041176</td>\n",
       "      <td>7.0</td>\n",
       "      <td>320.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>map00405</td>\n",
       "      <td>0.000303</td>\n",
       "      <td>0.014986</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon   kegg_id    pvalue    qvalue  precision    recall   f1score   TP  \\\n",
       "0         0  map00020  0.000345  0.083798   0.075472  0.105263  0.087912  4.0   \n",
       "1         5  map00525  0.000033  0.005310   0.100000  0.666667  0.173913  2.0   \n",
       "2         5  map01053  0.000066  0.005310   0.100000  0.500000  0.166667  2.0   \n",
       "3         5  map01130  0.000055  0.005310   0.350000  0.021875  0.041176  7.0   \n",
       "4         5  map00405  0.000303  0.014986   0.100000  0.250000  0.142857  2.0   \n",
       "\n",
       "   target_set_size  imodulon_size  \n",
       "0             38.0           53.0  \n",
       "1              3.0           20.0  \n",
       "2              4.0           20.0  \n",
       "3            320.0           20.0  \n",
       "4              8.0           20.0  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_pathway_enrich.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>M00011</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.003509</td>\n",
       "      <td>0.075472</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.109589</td>\n",
       "      <td>4.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>M00149</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.003509</td>\n",
       "      <td>0.056604</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>M00009</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>0.005907</td>\n",
       "      <td>0.075472</td>\n",
       "      <td>0.160000</td>\n",
       "      <td>0.102564</td>\n",
       "      <td>4.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>M00376</td>\n",
       "      <td>0.000185</td>\n",
       "      <td>0.012721</td>\n",
       "      <td>0.056604</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>3.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>M00173</td>\n",
       "      <td>0.000603</td>\n",
       "      <td>0.027655</td>\n",
       "      <td>0.056604</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>3.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon kegg_id    pvalue    qvalue  precision    recall   f1score   TP  \\\n",
       "0         0  M00011  0.000026  0.003509   0.075472  0.200000  0.109589  4.0   \n",
       "1         0  M00149  0.000024  0.003509   0.056604  0.428571  0.100000  3.0   \n",
       "2         0  M00009  0.000064  0.005907   0.075472  0.160000  0.102564  4.0   \n",
       "3         0  M00376  0.000185  0.012721   0.056604  0.230769  0.090909  3.0   \n",
       "4         0  M00173  0.000603  0.027655   0.056604  0.157895  0.083333  3.0   \n",
       "\n",
       "   target_set_size  imodulon_size  \n",
       "0             20.0           53.0  \n",
       "1              7.0           53.0  \n",
       "2             25.0           53.0  \n",
       "3             13.0           53.0  \n",
       "4             19.0           53.0  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_module_enrich.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert KEGG IDs to human-readable names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "88fc0bae543f451592e02dfd99719c94",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/192 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bad KEGG ID: map01130\n",
      "Bad KEGG ID: -\n",
      "Bad KEGG ID: -\n",
      "Bad KEGG ID: map01130\n",
      "Bad KEGG ID: map00281\n",
      "Bad KEGG ID: -\n",
      "Bad KEGG ID: map01130\n",
      "Bad KEGG ID: map00072\n",
      "Bad KEGG ID: map00281\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "be68f2ff54ec44bea76e72cc160c4790",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/124 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bad KEGG ID: M00254\n",
      "Bad KEGG ID: M00479\n",
      "Bad KEGG ID: M00258\n",
      "Bad KEGG ID: M00188\n",
      "Bad KEGG ID: M00207\n",
      "Bad KEGG ID: M00242\n",
      "Bad KEGG ID: M00244\n",
      "Bad KEGG ID: M00178\n",
      "Bad KEGG ID: M00240\n",
      "Bad KEGG ID: M00200\n",
      "Bad KEGG ID: M00207\n",
      "Bad KEGG ID: M00236\n",
      "Bad KEGG ID: M00435\n",
      "Bad KEGG ID: M00436\n",
      "Bad KEGG ID: M00239\n",
      "Bad KEGG ID: M00222\n",
      "Bad KEGG ID: M00178\n",
      "Bad KEGG ID: M00216\n",
      "Bad KEGG ID: M00460\n",
      "Bad KEGG ID: M00491\n",
      "Bad KEGG ID: M00454\n",
      "Bad KEGG ID: M00670\n",
      "Bad KEGG ID: M00210\n",
      "Bad KEGG ID: M00669\n",
      "Bad KEGG ID: M00323\n",
      "Bad KEGG ID: M00167\n",
      "Bad KEGG ID: M00212\n",
      "Bad KEGG ID: M00215\n",
      "Bad KEGG ID: M00323\n",
      "Bad KEGG ID: M00216\n",
      "Bad KEGG ID: M00166\n",
      "Bad KEGG ID: M00233\n",
      "Bad KEGG ID: M00239\n",
      "Bad KEGG ID: M00237\n",
      "Bad KEGG ID: M00240\n",
      "Bad KEGG ID: M00206\n",
      "Bad KEGG ID: M00207\n",
      "Bad KEGG ID: M00191\n",
      "Bad KEGG ID: M00439\n",
      "Bad KEGG ID: M00239\n",
      "Bad KEGG ID: M00178\n",
      "Bad KEGG ID: M00179\n",
      "Bad KEGG ID: M00183\n",
      "Bad KEGG ID: M00439\n",
      "Bad KEGG ID: M00239\n",
      "Bad KEGG ID: M00237\n",
      "Bad KEGG ID: M00236\n",
      "Bad KEGG ID: M00205\n",
      "Bad KEGG ID: M00236\n",
      "Bad KEGG ID: M00205\n",
      "Bad KEGG ID: M00237\n",
      "Bad KEGG ID: M00201\n",
      "Bad KEGG ID: M00233\n",
      "Bad KEGG ID: M00208\n",
      "Bad KEGG ID: M00207\n"
     ]
    }
   ],
   "source": [
    "import urllib.error\n",
    "\n",
    "for idx,key in tqdm(DF_pathway_enrich.kegg_id.items(),total=len(DF_pathway_enrich)):\n",
    "    try:\n",
    "        text = REST.kegg_find('pathway',key).read()\n",
    "        name = re.search('\\t(.*)\\n',text).group(1)\n",
    "        DF_pathway_enrich.loc[idx,'pathway_name'] = name\n",
    "    except (AttributeError, urllib.error.HTTPError):\n",
    "        print(f\"Bad KEGG ID: {key}\")\n",
    "        DF_pathway_enrich.loc[idx,'pathway_name'] = None\n",
    "    \n",
    "for idx,key in tqdm(DF_module_enrich.kegg_id.items(),total=len(DF_module_enrich)):\n",
    "    try:\n",
    "        text = REST.kegg_find('module',key).read()\n",
    "        name = re.search('\\t(.*)\\n',text).group(1)\n",
    "        DF_module_enrich.loc[idx,'module_name'] = name\n",
    "    except (AttributeError, urllib.error.HTTPError):\n",
    "        print(f\"Bad KEGG ID: {key}\")\n",
    "        DF_module_enrich.loc[idx,'module_name'] = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>pathway_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>map00020</td>\n",
       "      <td>0.000345</td>\n",
       "      <td>0.083798</td>\n",
       "      <td>0.075472</td>\n",
       "      <td>0.105263</td>\n",
       "      <td>0.087912</td>\n",
       "      <td>4.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>Citrate cycle (TCA cycle)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>map00525</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.005310</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.173913</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>Acarbose and validamycin biosynthesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>map01053</td>\n",
       "      <td>0.000066</td>\n",
       "      <td>0.005310</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>Biosynthesis of siderophore group nonribosomal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>map01130</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.005310</td>\n",
       "      <td>0.350000</td>\n",
       "      <td>0.021875</td>\n",
       "      <td>0.041176</td>\n",
       "      <td>7.0</td>\n",
       "      <td>320.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>map00405</td>\n",
       "      <td>0.000303</td>\n",
       "      <td>0.014986</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>Phenazine biosynthesis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon   kegg_id    pvalue    qvalue  precision    recall   f1score   TP  \\\n",
       "0         0  map00020  0.000345  0.083798   0.075472  0.105263  0.087912  4.0   \n",
       "1         5  map00525  0.000033  0.005310   0.100000  0.666667  0.173913  2.0   \n",
       "2         5  map01053  0.000066  0.005310   0.100000  0.500000  0.166667  2.0   \n",
       "3         5  map01130  0.000055  0.005310   0.350000  0.021875  0.041176  7.0   \n",
       "4         5  map00405  0.000303  0.014986   0.100000  0.250000  0.142857  2.0   \n",
       "\n",
       "   target_set_size  imodulon_size  \\\n",
       "0             38.0           53.0   \n",
       "1              3.0           20.0   \n",
       "2              4.0           20.0   \n",
       "3            320.0           20.0   \n",
       "4              8.0           20.0   \n",
       "\n",
       "                                        pathway_name  \n",
       "0                          Citrate cycle (TCA cycle)  \n",
       "1              Acarbose and validamycin biosynthesis  \n",
       "2  Biosynthesis of siderophore group nonribosomal...  \n",
       "3                                               None  \n",
       "4                             Phenazine biosynthesis  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_pathway_enrich.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>module_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>M00011</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.003509</td>\n",
       "      <td>0.075472</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>0.109589</td>\n",
       "      <td>4.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>Citrate cycle, second carbon oxidation, 2-oxog...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>M00149</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.003509</td>\n",
       "      <td>0.056604</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>Succinate dehydrogenase, prokaryotes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>M00009</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>0.005907</td>\n",
       "      <td>0.075472</td>\n",
       "      <td>0.160000</td>\n",
       "      <td>0.102564</td>\n",
       "      <td>4.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>Citrate cycle (TCA cycle, Krebs cycle)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>M00376</td>\n",
       "      <td>0.000185</td>\n",
       "      <td>0.012721</td>\n",
       "      <td>0.056604</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>3.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>3-Hydroxypropionate bi-cycle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>M00173</td>\n",
       "      <td>0.000603</td>\n",
       "      <td>0.027655</td>\n",
       "      <td>0.056604</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>3.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>Reductive citrate cycle (Arnon-Buchanan cycle)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon kegg_id    pvalue    qvalue  precision    recall   f1score   TP  \\\n",
       "0         0  M00011  0.000026  0.003509   0.075472  0.200000  0.109589  4.0   \n",
       "1         0  M00149  0.000024  0.003509   0.056604  0.428571  0.100000  3.0   \n",
       "2         0  M00009  0.000064  0.005907   0.075472  0.160000  0.102564  4.0   \n",
       "3         0  M00376  0.000185  0.012721   0.056604  0.230769  0.090909  3.0   \n",
       "4         0  M00173  0.000603  0.027655   0.056604  0.157895  0.083333  3.0   \n",
       "\n",
       "   target_set_size  imodulon_size  \\\n",
       "0             20.0           53.0   \n",
       "1              7.0           53.0   \n",
       "2             25.0           53.0   \n",
       "3             13.0           53.0   \n",
       "4             19.0           53.0   \n",
       "\n",
       "                                         module_name  \n",
       "0  Citrate cycle, second carbon oxidation, 2-oxog...  \n",
       "1               Succinate dehydrogenase, prokaryotes  \n",
       "2             Citrate cycle (TCA cycle, Krebs cycle)  \n",
       "3                       3-Hydroxypropionate bi-cycle  \n",
       "4     Reductive citrate cycle (Arnon-Buchanan cycle)  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_module_enrich.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO_enrich['source'] = 'GO'\n",
    "# DF_pathway_enrich['source'] = 'KEGG pathways'\n",
    "# DF_module_enrich['source'] = 'KEGG modules'\n",
    "# DF_subti_enrich['source'] = 'SubtiWiki'\n",
    "\n",
    "DF_GO_enrich.rename({'gene_ontology':'annotation'},axis=1, inplace=True)\n",
    "# DF_pathway_enrich.rename({'kegg_id':'annotation'},axis=1, inplace=True)\n",
    "# DF_module_enrich.rename({'kegg_id':'annotation'},axis=1, inplace=True)\n",
    "# DF_subti_enrich.rename({'value':'annotation'},axis=1, inplace=True)\n",
    "\n",
    "DF_enrichments = pd.concat([DF_GO_enrich, DF_pathway_enrich, DF_module_enrich])\n",
    "DF_enrichments.to_csv(path.join(data_dir,'functional_enrichments.csv'))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check for single gene iModulons"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some iModulons are dominated by a single, high-coefficient gene. These iModulons may result from:\n",
    "1. Overdecomposition of the dataset to identify noisy genes\n",
    "1. Artificial knock-out of single genes\n",
    "1. Regulons with only one target gene\n",
    "\n",
    "No matter what causes these iModulons, it is important to be aware of them. The find_single_gene_imodulons function identifies iModulons that are likely dominated by a single gene.\n",
    "\n",
    "The iModulons identified by ``find_single_gene_imodulons`` may contain more than one gene, since a threshold-agnostic method is used to identify these iModulons."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg_imods = ica_data.find_single_gene_imodulons(save=True)\n",
    "len(sg_imods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i,mod in enumerate(sg_imods):\n",
    "    ica_data.rename_imodulons({mod:'SG_'+str(i+1)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'single_gene'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/90/3j1z360d3ys7y_4lx77mk_br0000gp/T/ipykernel_29791/1381282101.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mica_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimodulon_table\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mica_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimodulon_table\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msingle_gene\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/imodulon/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5485\u001b[0m         ):\n\u001b[1;32m   5486\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5487\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5488\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5489\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'single_gene'"
     ]
    }
   ],
   "source": [
    "ica_data.imodulon_table[ica_data.imodulon_table.single_gene == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_weight</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>eggNOG_OGs</th>\n",
       "      <th>Description</th>\n",
       "      <th>GOs</th>\n",
       "      <th>EC</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>KEGG_Module</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>...</th>\n",
       "      <th>BiGG_Reaction</th>\n",
       "      <th>PFAMs</th>\n",
       "      <th>accession</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>COG</th>\n",
       "      <th>operon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>XNR_RS29710</th>\n",
       "      <td>0.091595</td>\n",
       "      <td>XNR_RS29710</td>\n",
       "      <td>COG2172@1|root,COG2172@2|Bacteria,2IMCZ@201174...</td>\n",
       "      <td>Histidine kinase-like ATPase domain</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>HATPase_c_2</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2376752</td>\n",
       "      <td>2377186</td>\n",
       "      <td>+</td>\n",
       "      <td>ATP-binding protein</td>\n",
       "      <td>Signal transduction mechanisms</td>\n",
       "      <td>Op1487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12550</th>\n",
       "      <td>0.169050</td>\n",
       "      <td>XNR_RS12550</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_2530</td>\n",
       "      <td>2910742</td>\n",
       "      <td>2910921</td>\n",
       "      <td>+</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>Op1883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12555</th>\n",
       "      <td>0.174173</td>\n",
       "      <td>XNR_RS12555</td>\n",
       "      <td>2C2M0@1|root,2ZTSU@2|Bacteria,2GR3G@201174|Act...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_2531</td>\n",
       "      <td>2911012</td>\n",
       "      <td>2911260</td>\n",
       "      <td>+</td>\n",
       "      <td>DUF6284 family protein</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>Op1884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12560</th>\n",
       "      <td>0.187321</td>\n",
       "      <td>XNR_RS12560</td>\n",
       "      <td>2DDGX@1|root,2ZI1K@2|Bacteria,2HSAS@201174|Act...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_2532</td>\n",
       "      <td>2911293</td>\n",
       "      <td>2912288</td>\n",
       "      <td>+</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>Op1884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12565</th>\n",
       "      <td>0.173812</td>\n",
       "      <td>XNR_RS12565</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2912285</td>\n",
       "      <td>2912615</td>\n",
       "      <td>+</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>Op1884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12570</th>\n",
       "      <td>0.172359</td>\n",
       "      <td>XNR_RS12570</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_2534</td>\n",
       "      <td>2912612</td>\n",
       "      <td>2912908</td>\n",
       "      <td>+</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>Op1884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12575</th>\n",
       "      <td>0.157271</td>\n",
       "      <td>XNR_RS12575</td>\n",
       "      <td>COG1051@1|root,COG1051@2|Bacteria,2IK42@201174...</td>\n",
       "      <td>Belongs to the Nudix hydrolase family</td>\n",
       "      <td>-</td>\n",
       "      <td>3.6.1.55</td>\n",
       "      <td>ko:K03574</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>ATP_bind_2,NUDIX</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_2535</td>\n",
       "      <td>2913048</td>\n",
       "      <td>2913446</td>\n",
       "      <td>+</td>\n",
       "      <td>NUDIX hydrolase</td>\n",
       "      <td>Nucleotide transport and metabolism</td>\n",
       "      <td>Op1885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12580</th>\n",
       "      <td>0.155842</td>\n",
       "      <td>XNR_RS12580</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_2536</td>\n",
       "      <td>2913449</td>\n",
       "      <td>2913664</td>\n",
       "      <td>+</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>Op1885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12595</th>\n",
       "      <td>0.169764</td>\n",
       "      <td>XNR_RS12595</td>\n",
       "      <td>2CIKG@1|root,2ZS5S@2|Bacteria,2H230@201174|Act...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_2539</td>\n",
       "      <td>2914615</td>\n",
       "      <td>2916195</td>\n",
       "      <td>+</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>Op1887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12600</th>\n",
       "      <td>0.150063</td>\n",
       "      <td>XNR_RS12600</td>\n",
       "      <td>COG1674@1|root,COG1674@2|Bacteria,2I9XY@201174...</td>\n",
       "      <td>ftsk spoiiie</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>FtsK_SpoIIIE</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_2540</td>\n",
       "      <td>2916195</td>\n",
       "      <td>2918303</td>\n",
       "      <td>+</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>Cell cycle control, cell division, chromosome ...</td>\n",
       "      <td>Op1887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12605</th>\n",
       "      <td>0.183991</td>\n",
       "      <td>XNR_RS12605</td>\n",
       "      <td>2A54S@1|root,30TTE@2|Bacteria,2IMCD@201174|Act...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_2541</td>\n",
       "      <td>2918531</td>\n",
       "      <td>2919010</td>\n",
       "      <td>+</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>Op1888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12610</th>\n",
       "      <td>0.133034</td>\n",
       "      <td>XNR_RS12610</td>\n",
       "      <td>2B0C3@1|root,31SP7@2|Bacteria,2GQZT@201174|Act...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_2542</td>\n",
       "      <td>2919027</td>\n",
       "      <td>2919215</td>\n",
       "      <td>+</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>Op1888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12615</th>\n",
       "      <td>0.185233</td>\n",
       "      <td>XNR_RS12615</td>\n",
       "      <td>28XD5@1|root,2ZJAW@2|Bacteria,2IQAK@201174|Act...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_2543</td>\n",
       "      <td>2919229</td>\n",
       "      <td>2919690</td>\n",
       "      <td>+</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>Op1888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12625</th>\n",
       "      <td>0.171688</td>\n",
       "      <td>XNR_RS12625</td>\n",
       "      <td>COG0270@1|root,COG0270@2|Bacteria,2GNB2@201174...</td>\n",
       "      <td>DNA (cytosine-5-)-methyltransferase activity</td>\n",
       "      <td>-</td>\n",
       "      <td>2.1.1.37</td>\n",
       "      <td>ko:K00558</td>\n",
       "      <td>ko00270,ko01100,ko05206,map00270,map01100,map0...</td>\n",
       "      <td>M00035</td>\n",
       "      <td>R04858</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2920816</td>\n",
       "      <td>2921508</td>\n",
       "      <td>+</td>\n",
       "      <td>DNA cytosine methyltransferase</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>Op1890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12630</th>\n",
       "      <td>0.167142</td>\n",
       "      <td>XNR_RS12630</td>\n",
       "      <td>COG0270@1|root,COG0270@2|Bacteria,2GNB2@201174...</td>\n",
       "      <td>DNA (cytosine-5-)-methyltransferase activity</td>\n",
       "      <td>-</td>\n",
       "      <td>2.1.1.37</td>\n",
       "      <td>ko:K00558</td>\n",
       "      <td>ko00270,ko01100,ko05206,map00270,map01100,map0...</td>\n",
       "      <td>M00035</td>\n",
       "      <td>R04858</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2921505</td>\n",
       "      <td>2922161</td>\n",
       "      <td>+</td>\n",
       "      <td>DNA cytosine methyltransferase</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>Op1890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12635</th>\n",
       "      <td>0.173126</td>\n",
       "      <td>XNR_RS12635</td>\n",
       "      <td>COG3378@1|root,COG3378@2|Bacteria,2I37I@201174...</td>\n",
       "      <td>Bifunctional DNA primase/polymerase, N-terminal</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Prim-Pol</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_2546</td>\n",
       "      <td>2922158</td>\n",
       "      <td>2923048</td>\n",
       "      <td>+</td>\n",
       "      <td>bifunctional DNA primase/polymerase</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>Op1890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12640</th>\n",
       "      <td>0.175533</td>\n",
       "      <td>XNR_RS12640</td>\n",
       "      <td>2B1DJ@1|root,31TU7@2|Bacteria,2GT0N@201174|Act...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_2547</td>\n",
       "      <td>2923045</td>\n",
       "      <td>2923395</td>\n",
       "      <td>+</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>Op1890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12645</th>\n",
       "      <td>0.148022</td>\n",
       "      <td>XNR_RS12645</td>\n",
       "      <td>COG0464@1|root,COG0464@2|Bacteria,2GJHA@201174...</td>\n",
       "      <td>Protein of unknown function (DUF3631)</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>DUF3631</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_2548</td>\n",
       "      <td>2923392</td>\n",
       "      <td>2924633</td>\n",
       "      <td>+</td>\n",
       "      <td>DUF3631 domain-containing protein</td>\n",
       "      <td>Posttranslational modification, protein turnov...</td>\n",
       "      <td>Op1890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12650</th>\n",
       "      <td>0.178520</td>\n",
       "      <td>XNR_RS12650</td>\n",
       "      <td>COG3311@1|root,COG3311@2|Bacteria,2GSCB@201174...</td>\n",
       "      <td>Helix-turn-helix domain</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>HTH_17</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_2549</td>\n",
       "      <td>2924807</td>\n",
       "      <td>2925016</td>\n",
       "      <td>+</td>\n",
       "      <td>helix-turn-helix domain-containing protein</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>Op1891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS12655</th>\n",
       "      <td>0.098659</td>\n",
       "      <td>XNR_RS12655</td>\n",
       "      <td>COG0582@1|root,COG0582@2|Bacteria,2GMMI@201174...</td>\n",
       "      <td>Belongs to the 'phage' integrase family</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Phage_int_SAM_3,Phage_int_SAM_5,Phage_integrase</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_2550</td>\n",
       "      <td>2925084</td>\n",
       "      <td>2926481</td>\n",
       "      <td>+</td>\n",
       "      <td>tyrosine-type recombinase/integrase</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>Op1892</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_weight    gene_name  \\\n",
       "XNR_RS29710     0.091595  XNR_RS29710   \n",
       "XNR_RS12550     0.169050  XNR_RS12550   \n",
       "XNR_RS12555     0.174173  XNR_RS12555   \n",
       "XNR_RS12560     0.187321  XNR_RS12560   \n",
       "XNR_RS12565     0.173812  XNR_RS12565   \n",
       "XNR_RS12570     0.172359  XNR_RS12570   \n",
       "XNR_RS12575     0.157271  XNR_RS12575   \n",
       "XNR_RS12580     0.155842  XNR_RS12580   \n",
       "XNR_RS12595     0.169764  XNR_RS12595   \n",
       "XNR_RS12600     0.150063  XNR_RS12600   \n",
       "XNR_RS12605     0.183991  XNR_RS12605   \n",
       "XNR_RS12610     0.133034  XNR_RS12610   \n",
       "XNR_RS12615     0.185233  XNR_RS12615   \n",
       "XNR_RS12625     0.171688  XNR_RS12625   \n",
       "XNR_RS12630     0.167142  XNR_RS12630   \n",
       "XNR_RS12635     0.173126  XNR_RS12635   \n",
       "XNR_RS12640     0.175533  XNR_RS12640   \n",
       "XNR_RS12645     0.148022  XNR_RS12645   \n",
       "XNR_RS12650     0.178520  XNR_RS12650   \n",
       "XNR_RS12655     0.098659  XNR_RS12655   \n",
       "\n",
       "                                                    eggNOG_OGs  \\\n",
       "XNR_RS29710  COG2172@1|root,COG2172@2|Bacteria,2IMCZ@201174...   \n",
       "XNR_RS12550                                                NaN   \n",
       "XNR_RS12555  2C2M0@1|root,2ZTSU@2|Bacteria,2GR3G@201174|Act...   \n",
       "XNR_RS12560  2DDGX@1|root,2ZI1K@2|Bacteria,2HSAS@201174|Act...   \n",
       "XNR_RS12565                                                NaN   \n",
       "XNR_RS12570                                                NaN   \n",
       "XNR_RS12575  COG1051@1|root,COG1051@2|Bacteria,2IK42@201174...   \n",
       "XNR_RS12580                                                NaN   \n",
       "XNR_RS12595  2CIKG@1|root,2ZS5S@2|Bacteria,2H230@201174|Act...   \n",
       "XNR_RS12600  COG1674@1|root,COG1674@2|Bacteria,2I9XY@201174...   \n",
       "XNR_RS12605  2A54S@1|root,30TTE@2|Bacteria,2IMCD@201174|Act...   \n",
       "XNR_RS12610  2B0C3@1|root,31SP7@2|Bacteria,2GQZT@201174|Act...   \n",
       "XNR_RS12615  28XD5@1|root,2ZJAW@2|Bacteria,2IQAK@201174|Act...   \n",
       "XNR_RS12625  COG0270@1|root,COG0270@2|Bacteria,2GNB2@201174...   \n",
       "XNR_RS12630  COG0270@1|root,COG0270@2|Bacteria,2GNB2@201174...   \n",
       "XNR_RS12635  COG3378@1|root,COG3378@2|Bacteria,2I37I@201174...   \n",
       "XNR_RS12640  2B1DJ@1|root,31TU7@2|Bacteria,2GT0N@201174|Act...   \n",
       "XNR_RS12645  COG0464@1|root,COG0464@2|Bacteria,2GJHA@201174...   \n",
       "XNR_RS12650  COG3311@1|root,COG3311@2|Bacteria,2GSCB@201174...   \n",
       "XNR_RS12655  COG0582@1|root,COG0582@2|Bacteria,2GMMI@201174...   \n",
       "\n",
       "                                                 Description  GOs        EC  \\\n",
       "XNR_RS29710              Histidine kinase-like ATPase domain    -         -   \n",
       "XNR_RS12550                                              NaN  NaN       NaN   \n",
       "XNR_RS12555                                                -    -         -   \n",
       "XNR_RS12560                                                -    -         -   \n",
       "XNR_RS12565                                              NaN  NaN       NaN   \n",
       "XNR_RS12570                                              NaN  NaN       NaN   \n",
       "XNR_RS12575            Belongs to the Nudix hydrolase family    -  3.6.1.55   \n",
       "XNR_RS12580                                              NaN  NaN       NaN   \n",
       "XNR_RS12595                                                -    -         -   \n",
       "XNR_RS12600                                     ftsk spoiiie    -         -   \n",
       "XNR_RS12605                                                -    -         -   \n",
       "XNR_RS12610                                                -    -         -   \n",
       "XNR_RS12615                                                -    -         -   \n",
       "XNR_RS12625     DNA (cytosine-5-)-methyltransferase activity    -  2.1.1.37   \n",
       "XNR_RS12630     DNA (cytosine-5-)-methyltransferase activity    -  2.1.1.37   \n",
       "XNR_RS12635  Bifunctional DNA primase/polymerase, N-terminal    -         -   \n",
       "XNR_RS12640                                                -    -         -   \n",
       "XNR_RS12645            Protein of unknown function (DUF3631)    -         -   \n",
       "XNR_RS12650                          Helix-turn-helix domain    -         -   \n",
       "XNR_RS12655          Belongs to the 'phage' integrase family    -         -   \n",
       "\n",
       "               KEGG_ko                                       KEGG_Pathway  \\\n",
       "XNR_RS29710          -                                                  -   \n",
       "XNR_RS12550        NaN                                                NaN   \n",
       "XNR_RS12555          -                                                  -   \n",
       "XNR_RS12560          -                                                  -   \n",
       "XNR_RS12565        NaN                                                NaN   \n",
       "XNR_RS12570        NaN                                                NaN   \n",
       "XNR_RS12575  ko:K03574                                                  -   \n",
       "XNR_RS12580        NaN                                                NaN   \n",
       "XNR_RS12595          -                                                  -   \n",
       "XNR_RS12600          -                                                  -   \n",
       "XNR_RS12605          -                                                  -   \n",
       "XNR_RS12610          -                                                  -   \n",
       "XNR_RS12615          -                                                  -   \n",
       "XNR_RS12625  ko:K00558  ko00270,ko01100,ko05206,map00270,map01100,map0...   \n",
       "XNR_RS12630  ko:K00558  ko00270,ko01100,ko05206,map00270,map01100,map0...   \n",
       "XNR_RS12635          -                                                  -   \n",
       "XNR_RS12640          -                                                  -   \n",
       "XNR_RS12645          -                                                  -   \n",
       "XNR_RS12650          -                                                  -   \n",
       "XNR_RS12655          -                                                  -   \n",
       "\n",
       "            KEGG_Module KEGG_Reaction  ... BiGG_Reaction  \\\n",
       "XNR_RS29710           -             -  ...             -   \n",
       "XNR_RS12550         NaN           NaN  ...           NaN   \n",
       "XNR_RS12555           -             -  ...             -   \n",
       "XNR_RS12560           -             -  ...             -   \n",
       "XNR_RS12565         NaN           NaN  ...           NaN   \n",
       "XNR_RS12570         NaN           NaN  ...           NaN   \n",
       "XNR_RS12575           -             -  ...             -   \n",
       "XNR_RS12580         NaN           NaN  ...           NaN   \n",
       "XNR_RS12595           -             -  ...             -   \n",
       "XNR_RS12600           -             -  ...             -   \n",
       "XNR_RS12605           -             -  ...             -   \n",
       "XNR_RS12610           -             -  ...             -   \n",
       "XNR_RS12615           -             -  ...             -   \n",
       "XNR_RS12625      M00035        R04858  ...             -   \n",
       "XNR_RS12630      M00035        R04858  ...             -   \n",
       "XNR_RS12635           -             -  ...             -   \n",
       "XNR_RS12640           -             -  ...             -   \n",
       "XNR_RS12645           -             -  ...             -   \n",
       "XNR_RS12650           -             -  ...             -   \n",
       "XNR_RS12655           -             -  ...             -   \n",
       "\n",
       "                                                       PFAMs    accession  \\\n",
       "XNR_RS29710                                      HATPase_c_2  NC_020990.1   \n",
       "XNR_RS12550                                              NaN  NC_020990.1   \n",
       "XNR_RS12555                                                -  NC_020990.1   \n",
       "XNR_RS12560                                                -  NC_020990.1   \n",
       "XNR_RS12565                                              NaN  NC_020990.1   \n",
       "XNR_RS12570                                              NaN  NC_020990.1   \n",
       "XNR_RS12575                                 ATP_bind_2,NUDIX  NC_020990.1   \n",
       "XNR_RS12580                                              NaN  NC_020990.1   \n",
       "XNR_RS12595                                                -  NC_020990.1   \n",
       "XNR_RS12600                                     FtsK_SpoIIIE  NC_020990.1   \n",
       "XNR_RS12605                                                -  NC_020990.1   \n",
       "XNR_RS12610                                                -  NC_020990.1   \n",
       "XNR_RS12615                                                -  NC_020990.1   \n",
       "XNR_RS12625                                                -  NC_020990.1   \n",
       "XNR_RS12630                                                -  NC_020990.1   \n",
       "XNR_RS12635                                         Prim-Pol  NC_020990.1   \n",
       "XNR_RS12640                                                -  NC_020990.1   \n",
       "XNR_RS12645                                          DUF3631  NC_020990.1   \n",
       "XNR_RS12650                                           HTH_17  NC_020990.1   \n",
       "XNR_RS12655  Phage_int_SAM_3,Phage_int_SAM_5,Phage_integrase  NC_020990.1   \n",
       "\n",
       "            old_locus_tag    start      end strand  \\\n",
       "XNR_RS29710           NaN  2376752  2377186      +   \n",
       "XNR_RS12550      XNR_2530  2910742  2910921      +   \n",
       "XNR_RS12555      XNR_2531  2911012  2911260      +   \n",
       "XNR_RS12560      XNR_2532  2911293  2912288      +   \n",
       "XNR_RS12565           NaN  2912285  2912615      +   \n",
       "XNR_RS12570      XNR_2534  2912612  2912908      +   \n",
       "XNR_RS12575      XNR_2535  2913048  2913446      +   \n",
       "XNR_RS12580      XNR_2536  2913449  2913664      +   \n",
       "XNR_RS12595      XNR_2539  2914615  2916195      +   \n",
       "XNR_RS12600      XNR_2540  2916195  2918303      +   \n",
       "XNR_RS12605      XNR_2541  2918531  2919010      +   \n",
       "XNR_RS12610      XNR_2542  2919027  2919215      +   \n",
       "XNR_RS12615      XNR_2543  2919229  2919690      +   \n",
       "XNR_RS12625           NaN  2920816  2921508      +   \n",
       "XNR_RS12630           NaN  2921505  2922161      +   \n",
       "XNR_RS12635      XNR_2546  2922158  2923048      +   \n",
       "XNR_RS12640      XNR_2547  2923045  2923395      +   \n",
       "XNR_RS12645      XNR_2548  2923392  2924633      +   \n",
       "XNR_RS12650      XNR_2549  2924807  2925016      +   \n",
       "XNR_RS12655      XNR_2550  2925084  2926481      +   \n",
       "\n",
       "                                           gene_product  \\\n",
       "XNR_RS29710                         ATP-binding protein   \n",
       "XNR_RS12550                        hypothetical protein   \n",
       "XNR_RS12555                      DUF6284 family protein   \n",
       "XNR_RS12560                        hypothetical protein   \n",
       "XNR_RS12565                        hypothetical protein   \n",
       "XNR_RS12570                        hypothetical protein   \n",
       "XNR_RS12575                             NUDIX hydrolase   \n",
       "XNR_RS12580                        hypothetical protein   \n",
       "XNR_RS12595                        hypothetical protein   \n",
       "XNR_RS12600                        hypothetical protein   \n",
       "XNR_RS12605                        hypothetical protein   \n",
       "XNR_RS12610                        hypothetical protein   \n",
       "XNR_RS12615                        hypothetical protein   \n",
       "XNR_RS12625              DNA cytosine methyltransferase   \n",
       "XNR_RS12630              DNA cytosine methyltransferase   \n",
       "XNR_RS12635         bifunctional DNA primase/polymerase   \n",
       "XNR_RS12640                        hypothetical protein   \n",
       "XNR_RS12645           DUF3631 domain-containing protein   \n",
       "XNR_RS12650  helix-turn-helix domain-containing protein   \n",
       "XNR_RS12655         tyrosine-type recombinase/integrase   \n",
       "\n",
       "                                                           COG  operon  \n",
       "XNR_RS29710                     Signal transduction mechanisms  Op1487  \n",
       "XNR_RS12550                                  No COG Annotation  Op1883  \n",
       "XNR_RS12555                                  No COG Annotation  Op1884  \n",
       "XNR_RS12560                                  No COG Annotation  Op1884  \n",
       "XNR_RS12565                                  No COG Annotation  Op1884  \n",
       "XNR_RS12570                                  No COG Annotation  Op1884  \n",
       "XNR_RS12575                Nucleotide transport and metabolism  Op1885  \n",
       "XNR_RS12580                                  No COG Annotation  Op1885  \n",
       "XNR_RS12595                                  No COG Annotation  Op1887  \n",
       "XNR_RS12600  Cell cycle control, cell division, chromosome ...  Op1887  \n",
       "XNR_RS12605                                  No COG Annotation  Op1888  \n",
       "XNR_RS12610                                  No COG Annotation  Op1888  \n",
       "XNR_RS12615                                  No COG Annotation  Op1888  \n",
       "XNR_RS12625              Replication, recombination and repair  Op1890  \n",
       "XNR_RS12630              Replication, recombination and repair  Op1890  \n",
       "XNR_RS12635                                   Function unknown  Op1890  \n",
       "XNR_RS12640                                  No COG Annotation  Op1890  \n",
       "XNR_RS12645  Posttranslational modification, protein turnov...  Op1890  \n",
       "XNR_RS12650                                      Transcription  Op1891  \n",
       "XNR_RS12655              Replication, recombination and repair  Op1892  \n",
       "\n",
       "[20 rows x 24 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.view_imodulon(12)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save iModulon object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.util import explained_variance\n",
    "from pymodulon.io import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add iModulon sizes and explained variance\n",
    "for im in ica_data.imodulon_names:\n",
    "    ica_data.imodulon_table.loc[im,'imodulon_size'] = len(ica_data.view_imodulon(im))\n",
    "    ica_data.imodulon_table.loc[im,'explained_variance'] = explained_variance(ica_data,imodulons=im)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>explained_variance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>53.0</td>\n",
       "      <td>0.026971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26.0</td>\n",
       "      <td>0.004914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20.0</td>\n",
       "      <td>0.001916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7.0</td>\n",
       "      <td>0.002046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4.0</td>\n",
       "      <td>0.002665</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon_size  explained_variance\n",
       "0           53.0            0.026971\n",
       "1           26.0            0.004914\n",
       "2           20.0            0.001916\n",
       "3            7.0            0.002046\n",
       "4            4.0            0.002665"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_table.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This will save your iModulon table, your thresholds, and any other information stored in the ica_data object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_to_json(ica_data, path.join('../..','data','interim','modulome','salb_raw.json.gz'))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you prefer to view and edit your iModulon table in excel, save it as a CSV and reload the iModulon as before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data.imodulon_table.to_csv(path.join('../..','data','interim','modulome','imodulon_table_raw.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the CSV file containing the imodulon names\n",
    "imodulon_names_df = pd.read_csv(path.join(interim_data, 'imodulon_table_raw.csv'))\n",
    "# Extract the names into a list\n",
    "imodulon_names = imodulon_names_df['Unnamed: 0'].tolist()\n",
    "all_genes = []  # List to store all genes from each imodulon\n",
    "\n",
    "for imodulon_name in imodulon_names:\n",
    "    imodulome_data = ica_data.view_imodulon(imodulon_name)\n",
    "    genes = imodulome_data['gene_name'].tolist()\n",
    "    for gene in genes:\n",
    "        all_genes.append((gene, imodulon_name))\n",
    "\n",
    "# Convert the collected gene data to a pandas DataFrame\n",
    "df = pd.DataFrame(all_genes, columns=['Gene', 'Imodulon'])\n",
    "\n",
    "# Save the DataFrame to a CSV file\n",
    "df.to_csv(path.join(data_dir,'genes_with_imodulon.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>explained_variance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>0.026971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.004914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.001916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.002046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.002665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>73</td>\n",
       "      <td>134.0</td>\n",
       "      <td>0.003439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>74</td>\n",
       "      <td>27.0</td>\n",
       "      <td>0.004961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>75</td>\n",
       "      <td>89.0</td>\n",
       "      <td>0.029237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>76</td>\n",
       "      <td>119.0</td>\n",
       "      <td>0.003165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>77</td>\n",
       "      <td>93.0</td>\n",
       "      <td>0.002598</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0  imodulon_size  explained_variance\n",
       "0            0           53.0            0.026971\n",
       "1            1           26.0            0.004914\n",
       "2            2           20.0            0.001916\n",
       "3            3            7.0            0.002046\n",
       "4            4            4.0            0.002665\n",
       "..         ...            ...                 ...\n",
       "73          73          134.0            0.003439\n",
       "74          74           27.0            0.004961\n",
       "75          75           89.0            0.029237\n",
       "76          76          119.0            0.003165\n",
       "77          77           93.0            0.002598\n",
       "\n",
       "[78 rows x 3 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imodulon_names_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
