{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.io import *\n",
    "from pymodulon.motif import *\n",
    "from os import path"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data = load_json_model(os.path.join('../..','data','processed','modulome','salb.json.gz'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_file = \"../../data/external/modulome/genome.fasta\""
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Search for motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read gene_nfo5.csv\n",
    "gene_info = pd.read_csv(\"../../data/processed/modulome/gene_info6.csv\",index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>eggNOG_OGs</th>\n",
       "      <th>Description</th>\n",
       "      <th>GOs</th>\n",
       "      <th>EC</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>KEGG_Module</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>KEGG_rclass</th>\n",
       "      <th>...</th>\n",
       "      <th>BiGG_Reaction</th>\n",
       "      <th>PFAMs</th>\n",
       "      <th>accession</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>COG</th>\n",
       "      <th>operon</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>XNR_RS30570</th>\n",
       "      <td>XNR_RS30570</td>\n",
       "      <td>COG1672@1|root,COG1672@2|Bacteria</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>AAA_22,TIR_2,TniB</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_0001</td>\n",
       "      <td>536</td>\n",
       "      <td>1609</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>Op0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS00010</th>\n",
       "      <td>XNR_RS00010</td>\n",
       "      <td>COG3935@1|root,COG3935@2|Bacteria,2IDRB@201174...</td>\n",
       "      <td>DnaD domain protein</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3324</td>\n",
       "      <td>3533</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>Replication, recombination and repair</td>\n",
       "      <td>Op2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS00015</th>\n",
       "      <td>XNR_RS00015</td>\n",
       "      <td>28PSB@1|root,2ZCDV@2|Bacteria,2GNDZ@201174|Act...</td>\n",
       "      <td>DDE superfamily endonuclease</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>DDE_Tnp_1,DDE_Tnp_4,HTH_Tnp_4</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_0003</td>\n",
       "      <td>3671</td>\n",
       "      <td>4441</td>\n",
       "      <td>-</td>\n",
       "      <td>IS5-like element IS112 family transposase</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>Op3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS00020</th>\n",
       "      <td>XNR_RS00020</td>\n",
       "      <td>2AKX1@1|root,31BQI@2|Bacteria,2GNMY@201174|Act...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_0004</td>\n",
       "      <td>4485</td>\n",
       "      <td>5069</td>\n",
       "      <td>-</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>No COG Annotation</td>\n",
       "      <td>Op4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS00025</th>\n",
       "      <td>XNR_RS00025</td>\n",
       "      <td>COG2944@1|root,COG2944@2|Bacteria,2GM0F@201174...</td>\n",
       "      <td>sequence-specific DNA binding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>DnaB_C,HTH_3,HTH_31</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_0005</td>\n",
       "      <td>5066</td>\n",
       "      <td>5983</td>\n",
       "      <td>-</td>\n",
       "      <td>dihydrolipoamide S-succinyltransferase</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>Op4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_name                                         eggNOG_OGs  \\\n",
       "gene_id                                                                       \n",
       "XNR_RS30570  XNR_RS30570                  COG1672@1|root,COG1672@2|Bacteria   \n",
       "XNR_RS00010  XNR_RS00010  COG3935@1|root,COG3935@2|Bacteria,2IDRB@201174...   \n",
       "XNR_RS00015  XNR_RS00015  28PSB@1|root,2ZCDV@2|Bacteria,2GNDZ@201174|Act...   \n",
       "XNR_RS00020  XNR_RS00020  2AKX1@1|root,31BQI@2|Bacteria,2GNMY@201174|Act...   \n",
       "XNR_RS00025  XNR_RS00025  COG2944@1|root,COG2944@2|Bacteria,2GM0F@201174...   \n",
       "\n",
       "                               Description GOs EC KEGG_ko KEGG_Pathway  \\\n",
       "gene_id                                                                  \n",
       "XNR_RS30570                              -   -  -       -            -   \n",
       "XNR_RS00010            DnaD domain protein   -  -       -            -   \n",
       "XNR_RS00015   DDE superfamily endonuclease   -  -       -            -   \n",
       "XNR_RS00020                              -   -  -       -            -   \n",
       "XNR_RS00025  sequence-specific DNA binding   -  -       -            -   \n",
       "\n",
       "            KEGG_Module KEGG_Reaction KEGG_rclass  ... BiGG_Reaction  \\\n",
       "gene_id                                            ...                 \n",
       "XNR_RS30570           -             -           -  ...             -   \n",
       "XNR_RS00010           -             -           -  ...             -   \n",
       "XNR_RS00015           -             -           -  ...             -   \n",
       "XNR_RS00020           -             -           -  ...             -   \n",
       "XNR_RS00025           -             -           -  ...             -   \n",
       "\n",
       "                                     PFAMs    accession old_locus_tag start  \\\n",
       "gene_id                                                                       \n",
       "XNR_RS30570              AAA_22,TIR_2,TniB  NC_020990.1      XNR_0001   536   \n",
       "XNR_RS00010                              -  NC_020990.1           NaN  3324   \n",
       "XNR_RS00015  DDE_Tnp_1,DDE_Tnp_4,HTH_Tnp_4  NC_020990.1      XNR_0003  3671   \n",
       "XNR_RS00020                              -  NC_020990.1      XNR_0004  4485   \n",
       "XNR_RS00025            DnaB_C,HTH_3,HTH_31  NC_020990.1      XNR_0005  5066   \n",
       "\n",
       "              end strand                               gene_product  \\\n",
       "gene_id                                                               \n",
       "XNR_RS30570  1609      -                       hypothetical protein   \n",
       "XNR_RS00010  3533      -                       hypothetical protein   \n",
       "XNR_RS00015  4441      -  IS5-like element IS112 family transposase   \n",
       "XNR_RS00020  5069      -                       hypothetical protein   \n",
       "XNR_RS00025  5983      -     dihydrolipoamide S-succinyltransferase   \n",
       "\n",
       "                                               COG operon  \n",
       "gene_id                                                    \n",
       "XNR_RS30570                      No COG Annotation    Op0  \n",
       "XNR_RS00010  Replication, recombination and repair    Op2  \n",
       "XNR_RS00015                       Function unknown    Op3  \n",
       "XNR_RS00020                      No COG Annotation    Op4  \n",
       "XNR_RS00025                          Transcription    Op4  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_info.head()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data.gene_table['gene_name'] = ica_data.gene_table['gene_name'].fillna('unknown')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for k in ica_data.imodulon_names:\n",
    "    try:\n",
    "        find_motifs(ica_data, k, fasta_file, force=True)\n",
    "    except ValueError as e:\n",
    "        print(f\"Skipping {k} due to error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BGC-19-20 deletion': <MotifInfo with 6 motifs across 159 sites>,\n",
       " 'Glutamine': <MotifInfo with 1 motif across 20 sites>,\n",
       " 'Prophages': <MotifInfo with 8 motifs across 53 sites>,\n",
       " 'Surugamide repressor': <MotifInfo with 3 motifs across 12 sites>,\n",
       " 'ArsR': <MotifInfo with 2 motifs across 4 sites>,\n",
       " 'Paulomycin-1': <MotifInfo with 10 motifs across 98 sites>,\n",
       " 'UC-1': <MotifInfo with 4 motifs across 430 sites>,\n",
       " 'Cyclofaulknamycin-1': <MotifInfo with 1 motif across 10 sites>,\n",
       " 'Leucine': <MotifInfo with 6 motifs across 36 sites>,\n",
       " 'UC-2': <MotifInfo with 4 motifs across 336 sites>,\n",
       " 'Histidine': <MotifInfo with 4 motifs across 25 sites>,\n",
       " 'UC-3': <MotifInfo with 5 motifs across 39 sites>,\n",
       " 'Cysteine': <MotifInfo with 2 motifs across 16 sites>,\n",
       " 'UC-4': <MotifInfo with 5 motifs across 392 sites>,\n",
       " 'Minimycin deletion': <MotifInfo with 7 motifs across 86 sites>,\n",
       " 'RpoH': <MotifInfo with 3 motifs across 120 sites>,\n",
       " 'Isorenieratene/streptamidine': <MotifInfo with 11 motifs across 89 sites>,\n",
       " 'WhiH': <MotifInfo with 5 motifs across 117 sites>,\n",
       " 'WblA': <MotifInfo with 2 motifs across 69 sites>,\n",
       " 'Sucrose': <MotifInfo with 3 motifs across 12 sites>,\n",
       " 'Sulfate': <MotifInfo with 2 motifs across 39 sites>,\n",
       " 'Phosphate': <MotifInfo with 4 motifs across 38 sites>,\n",
       " 'Ribosomal': <MotifInfo with 2 motifs across 59 sites>,\n",
       " 'Glycerol': <MotifInfo with 2 motifs across 25 sites>,\n",
       " 'RpoE': <MotifInfo with 3 motifs across 101 sites>,\n",
       " 'Phenylacetate': <MotifInfo with 1 motif across 2 sites>,\n",
       " 'WhiB3': <MotifInfo with 5 motifs across 278 sites>,\n",
       " 'Mce transporters': <MotifInfo with 2 motifs across 18 sites>,\n",
       " 'Nitrogen/Ammonia/Urea': <MotifInfo with 4 motifs across 27 sites>,\n",
       " 'Stress': <MotifInfo with 4 motifs across 265 sites>,\n",
       " 'Cytochrome': <MotifInfo with 3 motifs across 13 sites>,\n",
       " 'Transketolase': <MotifInfo with 3 motifs across 24 sites>,\n",
       " 'Carbohydrate metabolism-2': <MotifInfo with 6 motifs across 40 sites>,\n",
       " 'BGC-1 related': <MotifInfo with 5 motifs across 42 sites>,\n",
       " 'UC-6': <MotifInfo with 1 motif across 3 sites>,\n",
       " 'UC-7': <MotifInfo with 2 motifs across 91 sites>,\n",
       " 'NRPS BGC-14 related': <MotifInfo with 3 motifs across 58 sites>,\n",
       " 'Paulomycin-2': <MotifInfo with 8 motifs across 98 sites>,\n",
       " 'Minimycin/dudomycin': <MotifInfo with 3 motifs across 36 sites>,\n",
       " 'SigK': <MotifInfo with 2 motifs across 9 sites>,\n",
       " 'NDH': <MotifInfo with 4 motifs across 132 sites>,\n",
       " 'SAL-2242': <MotifInfo with 3 motifs across 12 sites>,\n",
       " 'LAP/Thiopeptide BGC-11': <MotifInfo with 3 motifs across 19 sites>,\n",
       " 'Cell membrane': <MotifInfo with 5 motifs across 293 sites>,\n",
       " 'Candicidin/antimycin': <MotifInfo with 8 motifs across 112 sites>,\n",
       " 'DNA damage': <MotifInfo with 6 motifs across 279 sites>,\n",
       " 'SigE': <MotifInfo with 2 motifs across 80 sites>,\n",
       " 'Glycosyltransferase': <MotifInfo with 1 motif across 8 sites>,\n",
       " 'Lipid metabolism': <MotifInfo with 5 motifs across 230 sites>,\n",
       " 'Large genome deletion': <MotifInfo with 6 motifs across 295 sites>,\n",
       " 'SGR PTMs': <MotifInfo with 8 motifs across 55 sites>,\n",
       " 'Iron acquisition': <MotifInfo with 9 motifs across 93 sites>,\n",
       " 'RpoS': <MotifInfo with 3 motifs across 110 sites>,\n",
       " 'UC-8': <MotifInfo with 2 motifs across 8 sites>,\n",
       " 'Cellobiose': <MotifInfo with 4 motifs across 40 sites>,\n",
       " 'Paulomycin-3': <MotifInfo with 6 motifs across 134 sites>,\n",
       " 'Fatty Acid': <MotifInfo with 5 motifs across 282 sites>,\n",
       " 'Nitrate': <MotifInfo with 4 motifs across 234 sites>,\n",
       " 'Nucleotide': <MotifInfo with 7 motifs across 48 sites>,\n",
       " 'Translation': <MotifInfo with 4 motifs across 242 sites>,\n",
       " 'Cell motility': <MotifInfo with 5 motifs across 144 sites>,\n",
       " 'Quorum sensing': <MotifInfo with 4 motifs across 307 sites>,\n",
       " 'Serine protease': <MotifInfo with 2 motifs across 43 sites>,\n",
       " 'Surugamide': <MotifInfo with 2 motifs across 111 sites>,\n",
       " 'Amino acid transport': <MotifInfo with 3 motifs across 269 sites>,\n",
       " 'UC-9': <MotifInfo with 3 motifs across 225 sites>}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.motif_info"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Convert .eps files to .png"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "from PIL import Image\n",
    "\n",
    "def convert_eps_to_png_in_folder(folder_path):\n",
    "    # Search for all .eps files in the folder and its subfolders\n",
    "    for eps_file_path in glob.glob(os.path.join(folder_path, '**/*.eps'), recursive=True):\n",
    "        # Create the .png file path by replacing the extension\n",
    "        png_file_path = os.path.splitext(eps_file_path)[0] + '.png'\n",
    "        \n",
    "        # Convert the .eps file to .png\n",
    "        img = Image.open(eps_file_path)\n",
    "        img.save(png_file_path, 'PNG')\n",
    "\n",
    "# Usage\n",
    "convert_eps_to_png_in_folder('motifs/')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare motifs to known databases using TOMTOM\n",
    "For some reason it doesn't work to just read the function from motifs.py, so I copied it here.\n",
    "\n",
    "We might want to adjust the thresholds and the databases used, according to other papers (See lab journal 17.01.23). The two papers using this method used slightly different parameters. I think the default is fine, as long as we state them in the methods section."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_motifs2(\n",
    "    motif_info: Optional[MotifInfo] = None,\n",
    "    motif_file: Optional[str] = None,\n",
    "    motif_db: Optional[str] = None,\n",
    "    outdir: Optional[str] = None,\n",
    "    force=False,\n",
    "    verbose=True,\n",
    "    evt=0.001,\n",
    "):\n",
    "    \"\"\"\n",
    "    Compare a MEME motif against external motif databases\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    motif_info: Optional[MotifInfo]\n",
    "        MotifInfo object. Either 'motif_info' or 'motif_file' must be provided.\n",
    "        'motif_info' takes precedence over 'motif_file'.\n",
    "    motif_file: Optional[os.PathLike]\n",
    "        Txt file generated by MEME from a motif search.\n",
    "    motif_db: Optional[Union[List,str]]\n",
    "        Name or path to MEME formatted databases\n",
    "    outdir: Optional[os.PathLike]\n",
    "        Output directory for TOMTOM comparisons\n",
    "    force: bool\n",
    "        Force execution of TOMTOM even if output already exists (default: False)\n",
    "    verbose: bool\n",
    "        Show steps in verbose output (default: True)\n",
    "    evt: float\n",
    "        E-value threshold (default: 0.001)\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    if motif_info:\n",
    "        motif_file = motif_info.file\n",
    "    elif motif_file is None:\n",
    "        raise ValueError(\"Either 'motif_info' or 'motif_file' must be provided\")\n",
    "\n",
    "    # Ensure motif file exists\n",
    "    if not os.path.isfile(motif_file):\n",
    "        raise FileNotFoundError(f\"File does not exist: '{motif_file}'\")\n",
    "\n",
    "    # Get motif DB locations\n",
    "    motif_db_dir = next(iter(pymodulon.data.meme.__path__))\n",
    "    default_dbs = [\n",
    "        \"collectf\",\n",
    "        \"dpinteract\",\n",
    "        \"prodoric\",\n",
    "        \"regtransbase\",\n",
    "        \"SwissRegulon_e_coli\",\n",
    "    ]\n",
    "\n",
    "    if motif_db is None:\n",
    "        motif_db = default_dbs\n",
    "        motif_db = [os.path.join(motif_db_dir, db + \".meme\") for db in motif_db]\n",
    "    elif isinstance(motif_db, str):\n",
    "        motif_db = [motif_db]\n",
    "\n",
    "    for db in motif_db:\n",
    "        if db in default_dbs:\n",
    "            motif_db = [os.path.join(motif_db_dir, db + \".meme\") for db in motif_db]\n",
    "        elif not os.path.isfile(db):\n",
    "            raise FileNotFoundError(f\"File does not exist: '{db}'\")\n",
    "\n",
    "    # Handle output directory\n",
    "    if outdir is None:\n",
    "        outdir = os.path.join(os.path.dirname(motif_file), \"tomtom_consensus\")\n",
    "\n",
    "    # Run TOMTOM\n",
    "    if not os.path.isdir(outdir) or force:\n",
    "        cmd = [\n",
    "            \"tomtom\",\n",
    "            \"-oc\",\n",
    "            outdir,\n",
    "            \"-thresh\",\n",
    "            str(evt),\n",
    "            \"-incomplete-scores\",\n",
    "            \"-png\",\n",
    "            motif_file,\n",
    "        ] + motif_db\n",
    "\n",
    "        if verbose:\n",
    "            print(\"Running command:\", \" \".join(cmd))\n",
    "\n",
    "        res = subprocess.check_call(cmd)\n",
    "        if res != 0:\n",
    "            cmd_str = \" \".join(cmd)\n",
    "            raise RuntimeError(\n",
    "                f\"TOMTOM failed to execute the following command: {cmd_str}\"\n",
    "            )\n",
    "    else result = _parse_tomtom(outdir)\n",
    "\n",
    "    # Save result\n",
    "    if motif_info:\n",
    "        motif_info.matches = result\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "for i in ica_data.motif_info:\n",
    "    try:\n",
    "        compare_motifs2(ica_data.motif_info[i], verbose=False)\n",
    "    except subprocess.CalledProcessError:\n",
    "        print(f\"An error occurred with {i}. Skipping...\")\n",
    "        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>rel_position</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>site_seq</th>\n",
       "      <th>genes</th>\n",
       "      <th>locus_tags</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>motif</th>\n",
       "      <th>operon</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">MEME-1</th>\n",
       "      <th>Op156</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>xylB</td>\n",
       "      <td>XNR_RS00955</td>\n",
       "      <td>233806</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Op1616</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XNR_RS10865,XNR_RS10870</td>\n",
       "      <td>XNR_RS10865,XNR_RS10870</td>\n",
       "      <td>2562222</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Op1617</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XNR_RS31180</td>\n",
       "      <td>XNR_RS31180</td>\n",
       "      <td>2562455</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Op1618</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XNR_RS31185</td>\n",
       "      <td>XNR_RS31185</td>\n",
       "      <td>2563259</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Op2596</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XNR_RS17100</td>\n",
       "      <td>XNR_RS17100</td>\n",
       "      <td>3932478</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">MEME-10</th>\n",
       "      <th>Op438</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XNR_RS02915</td>\n",
       "      <td>XNR_RS02915</td>\n",
       "      <td>696705</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Op439</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rfbA,rfbB</td>\n",
       "      <td>XNR_RS02920,XNR_RS02925</td>\n",
       "      <td>698918</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Op440</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XNR_RS02930,dhbA</td>\n",
       "      <td>XNR_RS02930,XNR_RS02935</td>\n",
       "      <td>699149</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Op441</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XNR_RS02940</td>\n",
       "      <td>XNR_RS02940</td>\n",
       "      <td>701380</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Op442</th>\n",
       "      <td>43</td>\n",
       "      <td>3.77e-10</td>\n",
       "      <td>CGTCGACGAGGTGGCTCCCGACCAGGAGAGCAGG</td>\n",
       "      <td>XNR_RS02945,XNR_RS02950</td>\n",
       "      <td>XNR_RS02945,XNR_RS02950</td>\n",
       "      <td>702159</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>210 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               rel_position    pvalue                            site_seq  \\\n",
       "motif   operon                                                              \n",
       "MEME-1  Op156           NaN       NaN                                 NaN   \n",
       "        Op1616          NaN       NaN                                 NaN   \n",
       "        Op1617          NaN       NaN                                 NaN   \n",
       "        Op1618          NaN       NaN                                 NaN   \n",
       "        Op2596          NaN       NaN                                 NaN   \n",
       "...                     ...       ...                                 ...   \n",
       "MEME-10 Op438           NaN       NaN                                 NaN   \n",
       "        Op439           NaN       NaN                                 NaN   \n",
       "        Op440           NaN       NaN                                 NaN   \n",
       "        Op441           NaN       NaN                                 NaN   \n",
       "        Op442            43  3.77e-10  CGTCGACGAGGTGGCTCCCGACCAGGAGAGCAGG   \n",
       "\n",
       "                                  genes               locus_tags  start_pos  \\\n",
       "motif   operon                                                                \n",
       "MEME-1  Op156                      xylB              XNR_RS00955     233806   \n",
       "        Op1616  XNR_RS10865,XNR_RS10870  XNR_RS10865,XNR_RS10870    2562222   \n",
       "        Op1617              XNR_RS31180              XNR_RS31180    2562455   \n",
       "        Op1618              XNR_RS31185              XNR_RS31185    2563259   \n",
       "        Op2596              XNR_RS17100              XNR_RS17100    3932478   \n",
       "...                                 ...                      ...        ...   \n",
       "MEME-10 Op438               XNR_RS02915              XNR_RS02915     696705   \n",
       "        Op439                 rfbA,rfbB  XNR_RS02920,XNR_RS02925     698918   \n",
       "        Op440          XNR_RS02930,dhbA  XNR_RS02930,XNR_RS02935     699149   \n",
       "        Op441               XNR_RS02940              XNR_RS02940     701380   \n",
       "        Op442   XNR_RS02945,XNR_RS02950  XNR_RS02945,XNR_RS02950     702159   \n",
       "\n",
       "               strand  \n",
       "motif   operon         \n",
       "MEME-1  Op156       +  \n",
       "        Op1616      -  \n",
       "        Op1617      -  \n",
       "        Op1618      -  \n",
       "        Op2596      -  \n",
       "...               ...  \n",
       "MEME-10 Op438       -  \n",
       "        Op439       -  \n",
       "        Op440       +  \n",
       "        Op441       +  \n",
       "        Op442       +  \n",
       "\n",
       "[210 rows x 7 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.motif_info['Paulomycin-1'].sites"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate a table summarizing results of motif search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "keys = list(ica_data.motif_info.keys())\n",
    "\n",
    "motif_table=pd.DataFrame(index=keys)\n",
    "\n",
    "for i in motif_table.index:\n",
    "    a = ica_data.motif_info[i].sites\n",
    "    \n",
    "    if i != 'PvdS':\n",
    "        motif_table.loc[i, 'Oplists'] = a.loc['MEME-1'][-a.loc['MEME-1','pvalue'].isna()].index\n",
    "    else:\n",
    "        motif_table.loc[i, 'Oplists'] = a.loc['MEME-4'][-a.loc['MEME-4','pvalue'].isna()].index\n",
    "\n",
    "for i in motif_table.index:\n",
    "    motif_table.loc[i,'Op_num']=len(motif_table.loc[i,'Oplists'])\n",
    "   \n",
    "    motif_table.loc[i,'Op_total_num']=len(ica_data.view_imodulon(i).operon.unique())\n",
    "\n",
    "\n",
    "    gene_num=0\n",
    "    \n",
    "    for j in motif_table.loc[i,'Oplists']:\n",
    "        gene_num = gene_num+len(ica_data.view_imodulon(i).operon[ica_data.view_imodulon(i).operon == j])\n",
    "        \n",
    "    motif_table.loc[i,'motif_gene_num'] = gene_num\n",
    "    \n",
    "    motif_table.loc[i,'gene_total_num']=len(ica_data.view_imodulon(i))\n",
    "    \n",
    "    if i != 'PvdS':\n",
    "        motif_table.loc[i,'consensus'] = ica_data.motif_info[i].motifs.loc['MEME-1','consensus']\n",
    "        motif_table.loc[i,'width'] = ica_data.motif_info[i].motifs.loc['MEME-1','width']\n",
    "\n",
    "    else:\n",
    "        motif_table.loc[i,'consensus'] = ica_data.motif_info[i].motifs.loc['MEME-4','consensus']\n",
    "        motif_table.loc[i,'width'] = ica_data.motif_info[i].motifs.loc['MEME-4','width']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Op_num</th>\n",
       "      <th>Op_total_num</th>\n",
       "      <th>motif_gene_num</th>\n",
       "      <th>gene_total_num</th>\n",
       "      <th>consensus</th>\n",
       "      <th>width</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BGC-19-20 deletion</th>\n",
       "      <td>42.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>WBSYCGWCGDCSNCGGYGACSWSSWSSDSSTSSWSGVMSD</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Glutamine</th>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>TSSYCRCCGHSSTCVCCGMSMKSSHCGVCSHSGSCSTS</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prophages</th>\n",
       "      <td>11.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>KCVYCCWCVRGHTCGACGGGGWSMTSSWCSRSYWCNTCG</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Surugamide repressor</th>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>CASVSCTTCRVHCTWKYTGAYATWCAACTASTAGAAC</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ArsR</th>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>TCTTGTTTCGACATGGGTCTATGTTGACAGCCTTCGATG</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Quorum sensing</th>\n",
       "      <td>105.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>131.0</td>\n",
       "      <td>134.0</td>\n",
       "      <td>KCSASGHCSNSSDCGNCGWSSWSSWSGVCSAGSTCS</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Serine protease</th>\n",
       "      <td>26.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>GCGGYSGSGGYGRSSGCGGBGGCGAGS</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Surugamide</th>\n",
       "      <td>62.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>GGYGVYSRSSRCGRBCGCSGCSRCGRCCGYSR</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Amino acid transport</th>\n",
       "      <td>107.0</td>\n",
       "      <td>107.0</td>\n",
       "      <td>119.0</td>\n",
       "      <td>119.0</td>\n",
       "      <td>HSSNSGHSVNGGMSGHSCDGSWSGBCSDSGMCSWSGHCSW</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UC-9</th>\n",
       "      <td>76.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>HCGNCGHSGMSSDSSWSGACSDSSNSGBCSKCSNCSRCSA</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>66 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Op_num  Op_total_num  motif_gene_num  gene_total_num  \\\n",
       "BGC-19-20 deletion      42.0          42.0            53.0            53.0   \n",
       "Glutamine               20.0          20.0            26.0            26.0   \n",
       "Prophages               11.0          12.0            19.0            20.0   \n",
       "Surugamide repressor     4.0           4.0             7.0             7.0   \n",
       "ArsR                     2.0           2.0             4.0             4.0   \n",
       "...                      ...           ...             ...             ...   \n",
       "Quorum sensing         105.0         108.0           131.0           134.0   \n",
       "Serine protease         26.0          26.0            27.0            27.0   \n",
       "Surugamide              62.0          62.0            89.0            89.0   \n",
       "Amino acid transport   107.0         107.0           119.0           119.0   \n",
       "UC-9                    76.0          78.0            91.0            93.0   \n",
       "\n",
       "                                                     consensus width  \n",
       "BGC-19-20 deletion    WBSYCGWCGDCSNCGGYGACSWSSWSSDSSTSSWSGVMSD    40  \n",
       "Glutamine               TSSYCRCCGHSSTCVCCGMSMKSSHCGVCSHSGSCSTS    38  \n",
       "Prophages              KCVYCCWCVRGHTCGACGGGGWSMTSSWCSRSYWCNTCG    39  \n",
       "Surugamide repressor     CASVSCTTCRVHCTWKYTGAYATWCAACTASTAGAAC    37  \n",
       "ArsR                   TCTTGTTTCGACATGGGTCTATGTTGACAGCCTTCGATG    39  \n",
       "...                                                        ...   ...  \n",
       "Quorum sensing            KCSASGHCSNSSDCGNCGWSSWSSWSGVCSAGSTCS    36  \n",
       "Serine protease                    GCGGYSGSGGYGRSSGCGGBGGCGAGS    27  \n",
       "Surugamide                    GGYGVYSRSSRCGRBCGCSGCSRCGRCCGYSR    32  \n",
       "Amino acid transport  HSSNSGHSVNGGMSGHSCDGSWSGBCSDSGMCSWSGHCSW    40  \n",
       "UC-9                  HCGNCGHSGMSSDSSWSGACSDSSNSGBCSKCSNCSRCSA    40  \n",
       "\n",
       "[66 rows x 6 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motif_table.drop(columns=['Oplists'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "motif_table.to_csv(\"../../data/processed/modulome/motif_table.csv\")         "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "66"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# how many rows are in the motif table?\n",
    "len(motif_table)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save the ica object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = path.join('../..','data','processed','modulome')\n",
    "save_to_json(ica_data, path.join(data_dir, 'salb.json.gz'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# convert .xml files for use with Difflogo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No 'MOTIF' line ending with 'MEME-1\\n' found in file ./motifs/Sugar-2/output.txt. Skipping this file.\n",
      "No 'MOTIF' line ending with 'MEME-1\\n' found in file ./motifs/Phospholipid/output.txt. Skipping this file.\n",
      "No 'MOTIF' line ending with 'MEME-1\\n' found in file ./motifs/Valine/output.txt. Skipping this file.\n",
      "No 'MOTIF' line ending with 'MEME-1\\n' found in file ./motifs/Potassium/output.txt. Skipping this file.\n",
      "No 'MOTIF' line ending with 'MEME-1\\n' found in file ./motifs/Fructose/output.txt. Skipping this file.\n",
      "No 'MOTIF' line ending with 'MEME-1\\n' found in file ./motifs/hrt_Complex/output.txt. Skipping this file.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "def read_pwm(file_path):\n",
    "    with open(file_path, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "\n",
    "    # Find the line that starts with \"MOTIF\" and ends with \"MEME-1\"\n",
    "    motif_line = next((i for i, line in enumerate(lines) if line.startswith('MOTIF') and line.endswith('MEME-1\\n')), None)\n",
    "\n",
    "    if motif_line is None:\n",
    "        print(f\"No 'MOTIF' line ending with 'MEME-1\\\\n' found in file {file_path}. Skipping this file.\")\n",
    "        return None\n",
    "\n",
    "    # Find the next line that starts with \"letter-probability matrix:\"\n",
    "    start_line = next(i for i, line in enumerate(lines[motif_line:]) if line.startswith('letter-probability matrix:'))\n",
    "\n",
    "    # Find the next line that starts with \"MOTIF\"\n",
    "    end_line = next((i for i, line in enumerate(lines[motif_line + start_line:]) if line.startswith('MOTIF')), None)\n",
    "\n",
    "    # If end_line is None, set it to the length of lines\n",
    "    if end_line is None:\n",
    "        end_line = len(lines) - motif_line - start_line\n",
    "\n",
    "    # Read the lines between start_line and end_line into a DataFrame\n",
    "    pwm = pd.DataFrame([line.split() for line in lines[motif_line + start_line + 1:motif_line + start_line + end_line]], dtype=float)\n",
    "\n",
    "    # Transpose the DataFrame\n",
    "    pwm = pwm.transpose()\n",
    "\n",
    "    return pwm\n",
    "\n",
    "# Get a list of all directories in './motifs/'\n",
    "directories = [d for d in os.listdir('./motifs/') if os.path.isdir(os.path.join('./motifs/', d))]\n",
    "\n",
    "# Process each directory\n",
    "for directory in directories:\n",
    "    # Use the function to read the PWM from the file\n",
    "    pwm = read_pwm(f'./motifs/{directory}/output.txt')\n",
    "    if pwm is None:\n",
    "        continue\n",
    "    # Remove columns with all NaN values\n",
    "    pwm = pwm.dropna(axis=1, how='all')\n",
    "    # Save the PWM to a file\n",
    "    pwm.to_csv(f'../../data/interim/modulome/pmws/{directory}_pwm.txt', sep='\\t', index=False, header=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
