{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup\n",
    "This IPython notebook will walk through the steps of characterizing iModulons through the semi-automated tools in PyModulon. You will need:\n",
    "\n",
    "* M and A matrices\n",
    "* Expression data (e.g. `log_tpm_norm.csv`)\n",
    "* Gene table and KEGG/GO annotations (Generated in `gene_annotation.ipynb`)\n",
    "* Sample table, with a column for `project` and `condition`\n",
    "* TRN file\n",
    "\n",
    "Optional:\n",
    "* iModulon table (if you already have some characterized iModulons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.core import IcaData\n",
    "from pymodulon.plotting import *\n",
    "from os import path\n",
    "import pandas as pd\n",
    "import re\n",
    "from Bio.KEGG import REST\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enter the location of your data here\n",
    "data_dir = path.join('../..','data','processed')\n",
    "\n",
    "# GO and KEGG annotations are in the 'external' folder\n",
    "external_data = path.join('../..','data','external')\n",
    "interim_data = path.join('../..','data','interim')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check your sample table (i.e. metadata file)\n",
    "Your metadata file will probably have a lot of columns, most of which you may not care about. Feel free to save a secondary copy of your metadata file with only columns that seem relevant to you. The two most important columns are:\n",
    "1. `project`\n",
    "2. `condition`\n",
    "\n",
    "Make sure that these columns exist in your metadata file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>experiment</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Control1-MSG</th>\n",
       "      <td>Control</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Control2-MSG</th>\n",
       "      <td>Control</td>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L-galactose-1</th>\n",
       "      <td>Azenta01</td>\n",
       "      <td>L-galactose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L-galactose-2</th>\n",
       "      <td>Azenta01</td>\n",
       "      <td>L-galactose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MSG-1A</th>\n",
       "      <td>Azenta01</td>\n",
       "      <td>Monosodium glutamate</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                project             condition\n",
       "experiment                                   \n",
       "Control1-MSG    Control               control\n",
       "Control2-MSG    Control               control\n",
       "L-galactose-1  Azenta01           L-galactose\n",
       "L-galactose-2  Azenta01           L-galactose\n",
       "MSG-1A         Azenta01  Monosodium glutamate"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata = pd.read_csv(path.join(data_dir,'metadata.tsv'),index_col=0,sep='\\t')\n",
    "df_metadata[['project','condition']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "print(df_metadata.project.notnull().all())\n",
    "print(df_metadata.condition.notnull().all())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Organize the data\n",
    "We need to modify the sample order of the A.csv file because it governs the order of samples in the dashboard. We will use the `df_metadata` to do this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read A.csv\n",
    "A = pd.read_csv(path.join(data_dir,'A.csv'),index_col=0)\n",
    "A = A.reindex(columns=df_metadata['sample_id'])\n",
    "A.to_csv(path.join(data_dir,'A_ordered.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read log_tpm_norm.csv\n",
    "log_tpm = pd.read_csv(path.join(data_dir,'log_tpm_norm.csv'),index_col=0)\n",
    "log_tpm = log_tpm.reindex(columns=df_metadata['sample_id'])\n",
    "log_tpm.to_csv(path.join(data_dir,'log_tpm_norm_ordered.csv'))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the data\n",
    "You're now ready to load your IcaData object!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n",
      "/Users/nilmat/miniconda3/envs/preprint/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "ica_data = IcaData(M = path.join(data_dir,'M.csv'),\n",
    "                   A = path.join(data_dir,'A_ordered.csv'),\n",
    "                   X = path.join(data_dir,'log_tpm_norm_ordered.csv'),\n",
    "                   gene_table = path.join(external_data,'gene_info.csv'),\n",
    "                   sample_table = path.join(data_dir,'metadata.tsv'),\n",
    "                   threshold_method='kmeans')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you don't have a TRN (or have a very minimal TRN), use `threshold_method = 'kmeans'`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ica_data = IcaData(M = path.join(data_dir,'M.csv'),\n",
    "#                    A = path.join(data_dir,'A.csv'),\n",
    "#                    X = path.join(data_dir,'log_tpm_norm.csv'),\n",
    "#                    gene_table = path.join(data_dir,'gene_info.csv'),\n",
    "#                    sample_table = path.join(data_dir,'metadata.tsv'),\n",
    "#                    trn = path.join(data_dir,'TRN.csv'),\n",
    "#                    threshold_method = 'kmeans')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functional iModulons"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GO annotations and KEGG pathways/modules were generated in the 1_create_the_gene_table.ipynb notebook. Enrichments will be calculated in this notebook, and further curated in the 3_manual_iModulon_curation notebook."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GO Enrichments"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First load the Gene Ontology annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>GOs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>XNR_RS00270</td>\n",
       "      <td>mannosyltransferase activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>XNR_RS00270</td>\n",
       "      <td>molecular_function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>XNR_RS00270</td>\n",
       "      <td>catalytic activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>XNR_RS00270</td>\n",
       "      <td>dolichyl-phosphate beta-D-mannosyltransferase ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>XNR_RS00270</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      locus_tag                                                GOs\n",
       "54  XNR_RS00270                       mannosyltransferase activity\n",
       "55  XNR_RS00270                                 molecular_function\n",
       "56  XNR_RS00270                                 catalytic activity\n",
       "57  XNR_RS00270  dolichyl-phosphate beta-D-mannosyltransferase ...\n",
       "58  XNR_RS00270                                 cellular_component"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_GO = pd.read_csv(path.join(external_data,'GO_annotations_curated.csv'),index_col=0)\n",
    "DF_GO.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change 'locus_tag' to 'gene_id'\n",
    "DF_GO = DF_GO.rename(columns={'locus_tag':'gene_id'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO_enrich = ica_data.compute_annotation_enrichment(DF_GO,'GOs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>GOs</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15</td>\n",
       "      <td>ribonucleoprotein complex</td>\n",
       "      <td>1.176132e-63</td>\n",
       "      <td>1.336674e-60</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.893617</td>\n",
       "      <td>0.417910</td>\n",
       "      <td>42.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>154.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15</td>\n",
       "      <td>ribosome</td>\n",
       "      <td>1.176132e-63</td>\n",
       "      <td>1.336674e-60</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.893617</td>\n",
       "      <td>0.417910</td>\n",
       "      <td>42.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>154.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>intracellular non-membrane-bounded organelle</td>\n",
       "      <td>3.889383e-61</td>\n",
       "      <td>1.768113e-58</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.840000</td>\n",
       "      <td>0.411765</td>\n",
       "      <td>42.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>154.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15</td>\n",
       "      <td>ribosomal subunit</td>\n",
       "      <td>2.810101e-61</td>\n",
       "      <td>1.768113e-58</td>\n",
       "      <td>0.259740</td>\n",
       "      <td>0.909091</td>\n",
       "      <td>0.404040</td>\n",
       "      <td>40.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>154.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15</td>\n",
       "      <td>non-membrane-bounded organelle</td>\n",
       "      <td>3.889383e-61</td>\n",
       "      <td>1.768113e-58</td>\n",
       "      <td>0.272727</td>\n",
       "      <td>0.840000</td>\n",
       "      <td>0.411765</td>\n",
       "      <td>42.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>154.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon                                           GOs        pvalue  \\\n",
       "0        15                     ribonucleoprotein complex  1.176132e-63   \n",
       "1        15                                      ribosome  1.176132e-63   \n",
       "2        15  intracellular non-membrane-bounded organelle  3.889383e-61   \n",
       "3        15                             ribosomal subunit  2.810101e-61   \n",
       "4        15                non-membrane-bounded organelle  3.889383e-61   \n",
       "\n",
       "         qvalue  precision    recall   f1score    TP  target_set_size  \\\n",
       "0  1.336674e-60   0.272727  0.893617  0.417910  42.0             47.0   \n",
       "1  1.336674e-60   0.272727  0.893617  0.417910  42.0             47.0   \n",
       "2  1.768113e-58   0.272727  0.840000  0.411765  42.0             50.0   \n",
       "3  1.768113e-58   0.259740  0.909091  0.404040  40.0             44.0   \n",
       "4  1.768113e-58   0.272727  0.840000  0.411765  42.0             50.0   \n",
       "\n",
       "   imodulon_size  \n",
       "0          154.0  \n",
       "1          154.0  \n",
       "2          154.0  \n",
       "3          154.0  \n",
       "4          154.0  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_GO_enrich.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## KEGG Enrichments"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load KEGG mapping\n",
    "The `kegg_mapping.csv` file contains KEGG orthologies, pathways, modules, and reactions. Only pathways and modules are relevant to iModulon characterization."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['KEGG_ko' 'KEGG_Pathway' 'KEGG_Module' 'KEGG_Reaction']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>database</th>\n",
       "      <th>kegg_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XNR_RS30570</td>\n",
       "      <td>KEGG_ko</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XNR_RS00010</td>\n",
       "      <td>KEGG_ko</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XNR_RS00015</td>\n",
       "      <td>KEGG_ko</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XNR_RS00020</td>\n",
       "      <td>KEGG_ko</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XNR_RS00025</td>\n",
       "      <td>KEGG_ko</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_id database kegg_id\n",
       "0  XNR_RS30570  KEGG_ko       -\n",
       "1  XNR_RS00010  KEGG_ko       -\n",
       "2  XNR_RS00015  KEGG_ko       -\n",
       "3  XNR_RS00020  KEGG_ko       -\n",
       "4  XNR_RS00025  KEGG_ko       -"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_KEGG = pd.read_csv(path.join(external_data,'kegg_mapping.csv'),index_col=0)\n",
    "print(DF_KEGG.database.unique())\n",
    "DF_KEGG.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg_pathways = DF_KEGG[DF_KEGG.database == 'KEGG_Pathway']\n",
    "kegg_modules = DF_KEGG[DF_KEGG.database == 'KEGG_Module']"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform enrichment\n",
    "Uses the `compute_annotation_enrichment` function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_pathway_enrich = ica_data.compute_annotation_enrichment(kegg_pathways,'kegg_id')\n",
    "DF_module_enrich = ica_data.compute_annotation_enrichment(kegg_modules,'kegg_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>map01212</td>\n",
       "      <td>3.272854e-26</td>\n",
       "      <td>7.953036e-24</td>\n",
       "      <td>0.216495</td>\n",
       "      <td>0.446809</td>\n",
       "      <td>0.291667</td>\n",
       "      <td>21.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>map01200</td>\n",
       "      <td>9.216777e-17</td>\n",
       "      <td>1.119838e-14</td>\n",
       "      <td>0.237113</td>\n",
       "      <td>0.153333</td>\n",
       "      <td>0.186235</td>\n",
       "      <td>23.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>map00071</td>\n",
       "      <td>1.505542e-15</td>\n",
       "      <td>9.756987e-14</td>\n",
       "      <td>0.144330</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.201439</td>\n",
       "      <td>14.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>map00280</td>\n",
       "      <td>1.606088e-15</td>\n",
       "      <td>9.756987e-14</td>\n",
       "      <td>0.154639</td>\n",
       "      <td>0.288462</td>\n",
       "      <td>0.201342</td>\n",
       "      <td>15.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>map00640</td>\n",
       "      <td>9.103613e-15</td>\n",
       "      <td>4.424356e-13</td>\n",
       "      <td>0.144330</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>0.194444</td>\n",
       "      <td>14.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon   kegg_id        pvalue        qvalue  precision    recall  \\\n",
       "0         0  map01212  3.272854e-26  7.953036e-24   0.216495  0.446809   \n",
       "1         0  map01200  9.216777e-17  1.119838e-14   0.237113  0.153333   \n",
       "2         0  map00071  1.505542e-15  9.756987e-14   0.144330  0.333333   \n",
       "3         0  map00280  1.606088e-15  9.756987e-14   0.154639  0.288462   \n",
       "4         0  map00640  9.103613e-15  4.424356e-13   0.144330  0.297872   \n",
       "\n",
       "    f1score    TP  target_set_size  imodulon_size  \n",
       "0  0.291667  21.0             47.0           97.0  \n",
       "1  0.186235  23.0            150.0           97.0  \n",
       "2  0.201439  14.0             42.0           97.0  \n",
       "3  0.201342  15.0             52.0           97.0  \n",
       "4  0.194444  14.0             47.0           97.0  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_pathway_enrich.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>M00373</td>\n",
       "      <td>6.401638e-13</td>\n",
       "      <td>1.760450e-10</td>\n",
       "      <td>0.092784</td>\n",
       "      <td>0.562500</td>\n",
       "      <td>0.159292</td>\n",
       "      <td>9.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>M00087</td>\n",
       "      <td>8.745400e-11</td>\n",
       "      <td>1.202493e-08</td>\n",
       "      <td>0.082474</td>\n",
       "      <td>0.470588</td>\n",
       "      <td>0.140351</td>\n",
       "      <td>8.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>M00088</td>\n",
       "      <td>1.384311e-09</td>\n",
       "      <td>1.268951e-07</td>\n",
       "      <td>0.061856</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.113208</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>M00375</td>\n",
       "      <td>3.414887e-09</td>\n",
       "      <td>2.347735e-07</td>\n",
       "      <td>0.061856</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.112150</td>\n",
       "      <td>6.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>M00095</td>\n",
       "      <td>5.903501e-08</td>\n",
       "      <td>3.246926e-06</td>\n",
       "      <td>0.051546</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>5.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon kegg_id        pvalue        qvalue  precision    recall  \\\n",
       "0         0  M00373  6.401638e-13  1.760450e-10   0.092784  0.562500   \n",
       "1         0  M00087  8.745400e-11  1.202493e-08   0.082474  0.470588   \n",
       "2         0  M00088  1.384311e-09  1.268951e-07   0.061856  0.666667   \n",
       "3         0  M00375  3.414887e-09  2.347735e-07   0.061856  0.600000   \n",
       "4         0  M00095  5.903501e-08  3.246926e-06   0.051546  0.625000   \n",
       "\n",
       "    f1score   TP  target_set_size  imodulon_size  \n",
       "0  0.159292  9.0             16.0           97.0  \n",
       "1  0.140351  8.0             17.0           97.0  \n",
       "2  0.113208  6.0              9.0           97.0  \n",
       "3  0.112150  6.0             10.0           97.0  \n",
       "4  0.095238  5.0              8.0           97.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_module_enrich.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert KEGG IDs to human-readable names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0b25d0974ad44ae9a8583e1e8638e9bc",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/163 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bad KEGG ID: map00072\n",
      "Bad KEGG ID: map00281\n",
      "Bad KEGG ID: map01130\n",
      "Bad KEGG ID: map01130\n",
      "Bad KEGG ID: map00281\n",
      "Bad KEGG ID: -\n",
      "Bad KEGG ID: -\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b122db7e0f0e4d5d90717c99b1e523cb",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/108 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bad KEGG ID: M00167\n",
      "Bad KEGG ID: M00323\n",
      "Bad KEGG ID: M00670\n",
      "Bad KEGG ID: M00210\n",
      "Bad KEGG ID: M00669\n",
      "Bad KEGG ID: M00178\n",
      "Bad KEGG ID: M00179\n",
      "Bad KEGG ID: M00183\n",
      "Bad KEGG ID: M00215\n",
      "Bad KEGG ID: M00208\n",
      "Bad KEGG ID: M00491\n",
      "Bad KEGG ID: M00206\n",
      "Bad KEGG ID: M00207\n",
      "Bad KEGG ID: M00240\n",
      "Bad KEGG ID: M00188\n",
      "Bad KEGG ID: -\n",
      "Bad KEGG ID: M00236\n",
      "Bad KEGG ID: M00435\n",
      "Bad KEGG ID: M00436\n",
      "Bad KEGG ID: M00239\n",
      "Bad KEGG ID: M00239\n",
      "Bad KEGG ID: M00237\n",
      "Bad KEGG ID: M00236\n",
      "Bad KEGG ID: M00439\n",
      "Bad KEGG ID: M00205\n",
      "Bad KEGG ID: M00201\n",
      "Bad KEGG ID: M00237\n",
      "Bad KEGG ID: M00200\n",
      "Bad KEGG ID: M00215\n",
      "Bad KEGG ID: M00216\n",
      "Bad KEGG ID: M00619\n",
      "Bad KEGG ID: M00178\n",
      "Bad KEGG ID: M00233\n",
      "Bad KEGG ID: M00208\n",
      "Bad KEGG ID: -\n",
      "Bad KEGG ID: M00460\n",
      "Bad KEGG ID: M00237\n",
      "Bad KEGG ID: M00439\n",
      "Bad KEGG ID: M00239\n",
      "Bad KEGG ID: M00222\n",
      "Bad KEGG ID: M00254\n",
      "Bad KEGG ID: M00216\n",
      "Bad KEGG ID: M00212\n",
      "Bad KEGG ID: M00454\n",
      "Bad KEGG ID: M00221\n",
      "Bad KEGG ID: M00335\n",
      "Bad KEGG ID: M00237\n",
      "Bad KEGG ID: M00254\n",
      "Bad KEGG ID: M00606\n",
      "Bad KEGG ID: M00239\n",
      "Bad KEGG ID: M00479\n",
      "Bad KEGG ID: M00254\n",
      "Bad KEGG ID: -\n",
      "Bad KEGG ID: M00207\n",
      "Bad KEGG ID: M00258\n"
     ]
    }
   ],
   "source": [
    "import urllib.error\n",
    "\n",
    "for idx,key in tqdm(DF_pathway_enrich.kegg_id.items(),total=len(DF_pathway_enrich)):\n",
    "    try:\n",
    "        text = REST.kegg_find('pathway',key).read()\n",
    "        name = re.search('\\t(.*)\\n',text).group(1)\n",
    "        DF_pathway_enrich.loc[idx,'pathway_name'] = name\n",
    "    except (AttributeError, urllib.error.HTTPError):\n",
    "        print(f\"Bad KEGG ID: {key}\")\n",
    "        DF_pathway_enrich.loc[idx,'pathway_name'] = None\n",
    "    \n",
    "for idx,key in tqdm(DF_module_enrich.kegg_id.items(),total=len(DF_module_enrich)):\n",
    "    try:\n",
    "        text = REST.kegg_find('module',key).read()\n",
    "        name = re.search('\\t(.*)\\n',text).group(1)\n",
    "        DF_module_enrich.loc[idx,'module_name'] = name\n",
    "    except (AttributeError, urllib.error.HTTPError):\n",
    "        print(f\"Bad KEGG ID: {key}\")\n",
    "        DF_module_enrich.loc[idx,'module_name'] = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>pathway_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>map01212</td>\n",
       "      <td>3.272854e-26</td>\n",
       "      <td>7.953036e-24</td>\n",
       "      <td>0.216495</td>\n",
       "      <td>0.446809</td>\n",
       "      <td>0.291667</td>\n",
       "      <td>21.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>Fatty acid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>map01200</td>\n",
       "      <td>9.216777e-17</td>\n",
       "      <td>1.119838e-14</td>\n",
       "      <td>0.237113</td>\n",
       "      <td>0.153333</td>\n",
       "      <td>0.186235</td>\n",
       "      <td>23.0</td>\n",
       "      <td>150.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>Carbon metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>map00071</td>\n",
       "      <td>1.505542e-15</td>\n",
       "      <td>9.756987e-14</td>\n",
       "      <td>0.144330</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.201439</td>\n",
       "      <td>14.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>Fatty acid degradation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>map00280</td>\n",
       "      <td>1.606088e-15</td>\n",
       "      <td>9.756987e-14</td>\n",
       "      <td>0.154639</td>\n",
       "      <td>0.288462</td>\n",
       "      <td>0.201342</td>\n",
       "      <td>15.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>Valine, leucine and isoleucine degradation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>map00640</td>\n",
       "      <td>9.103613e-15</td>\n",
       "      <td>4.424356e-13</td>\n",
       "      <td>0.144330</td>\n",
       "      <td>0.297872</td>\n",
       "      <td>0.194444</td>\n",
       "      <td>14.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>Propanoate metabolism</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon   kegg_id        pvalue        qvalue  precision    recall  \\\n",
       "0         0  map01212  3.272854e-26  7.953036e-24   0.216495  0.446809   \n",
       "1         0  map01200  9.216777e-17  1.119838e-14   0.237113  0.153333   \n",
       "2         0  map00071  1.505542e-15  9.756987e-14   0.144330  0.333333   \n",
       "3         0  map00280  1.606088e-15  9.756987e-14   0.154639  0.288462   \n",
       "4         0  map00640  9.103613e-15  4.424356e-13   0.144330  0.297872   \n",
       "\n",
       "    f1score    TP  target_set_size  imodulon_size  \\\n",
       "0  0.291667  21.0             47.0           97.0   \n",
       "1  0.186235  23.0            150.0           97.0   \n",
       "2  0.201439  14.0             42.0           97.0   \n",
       "3  0.201342  15.0             52.0           97.0   \n",
       "4  0.194444  14.0             47.0           97.0   \n",
       "\n",
       "                                 pathway_name  \n",
       "0                       Fatty acid metabolism  \n",
       "1                           Carbon metabolism  \n",
       "2                      Fatty acid degradation  \n",
       "3  Valine, leucine and isoleucine degradation  \n",
       "4                       Propanoate metabolism  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_pathway_enrich.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon</th>\n",
       "      <th>kegg_id</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>qvalue</th>\n",
       "      <th>precision</th>\n",
       "      <th>recall</th>\n",
       "      <th>f1score</th>\n",
       "      <th>TP</th>\n",
       "      <th>target_set_size</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>module_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>M00373</td>\n",
       "      <td>6.401638e-13</td>\n",
       "      <td>1.760450e-10</td>\n",
       "      <td>0.092784</td>\n",
       "      <td>0.562500</td>\n",
       "      <td>0.159292</td>\n",
       "      <td>9.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>Ethylmalonyl pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>M00087</td>\n",
       "      <td>8.745400e-11</td>\n",
       "      <td>1.202493e-08</td>\n",
       "      <td>0.082474</td>\n",
       "      <td>0.470588</td>\n",
       "      <td>0.140351</td>\n",
       "      <td>8.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>beta-Oxidation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>M00088</td>\n",
       "      <td>1.384311e-09</td>\n",
       "      <td>1.268951e-07</td>\n",
       "      <td>0.061856</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.113208</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>Ketone body biosynthesis, acetyl-CoA =&gt; acetoa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>M00375</td>\n",
       "      <td>3.414887e-09</td>\n",
       "      <td>2.347735e-07</td>\n",
       "      <td>0.061856</td>\n",
       "      <td>0.600000</td>\n",
       "      <td>0.112150</td>\n",
       "      <td>6.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>Hydroxypropionate-hydroxybutylate cycle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>M00095</td>\n",
       "      <td>5.903501e-08</td>\n",
       "      <td>3.246926e-06</td>\n",
       "      <td>0.051546</td>\n",
       "      <td>0.625000</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>5.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>C5 isoprenoid biosynthesis, mevalonate pathway</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon kegg_id        pvalue        qvalue  precision    recall  \\\n",
       "0         0  M00373  6.401638e-13  1.760450e-10   0.092784  0.562500   \n",
       "1         0  M00087  8.745400e-11  1.202493e-08   0.082474  0.470588   \n",
       "2         0  M00088  1.384311e-09  1.268951e-07   0.061856  0.666667   \n",
       "3         0  M00375  3.414887e-09  2.347735e-07   0.061856  0.600000   \n",
       "4         0  M00095  5.903501e-08  3.246926e-06   0.051546  0.625000   \n",
       "\n",
       "    f1score   TP  target_set_size  imodulon_size  \\\n",
       "0  0.159292  9.0             16.0           97.0   \n",
       "1  0.140351  8.0             17.0           97.0   \n",
       "2  0.113208  6.0              9.0           97.0   \n",
       "3  0.112150  6.0             10.0           97.0   \n",
       "4  0.095238  5.0              8.0           97.0   \n",
       "\n",
       "                                         module_name  \n",
       "0                               Ethylmalonyl pathway  \n",
       "1                                     beta-Oxidation  \n",
       "2  Ketone body biosynthesis, acetyl-CoA => acetoa...  \n",
       "3            Hydroxypropionate-hydroxybutylate cycle  \n",
       "4     C5 isoprenoid biosynthesis, mevalonate pathway  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_module_enrich.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO_enrich['source'] = 'GO'\n",
    "# DF_pathway_enrich['source'] = 'KEGG pathways'\n",
    "# DF_module_enrich['source'] = 'KEGG modules'\n",
    "# DF_subti_enrich['source'] = 'SubtiWiki'\n",
    "\n",
    "DF_GO_enrich.rename({'gene_ontology':'annotation'},axis=1, inplace=True)\n",
    "# DF_pathway_enrich.rename({'kegg_id':'annotation'},axis=1, inplace=True)\n",
    "# DF_module_enrich.rename({'kegg_id':'annotation'},axis=1, inplace=True)\n",
    "# DF_subti_enrich.rename({'value':'annotation'},axis=1, inplace=True)\n",
    "\n",
    "DF_enrichments = pd.concat([DF_GO_enrich, DF_pathway_enrich, DF_module_enrich])\n",
    "DF_enrichments.to_csv(path.join(data_dir,'functional_enrichments.csv'))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check for single gene iModulons"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some iModulons are dominated by a single, high-coefficient gene. These iModulons may result from:\n",
    "1. Overdecomposition of the dataset to identify noisy genes\n",
    "1. Artificial knock-out of single genes\n",
    "1. Regulons with only one target gene\n",
    "\n",
    "No matter what causes these iModulons, it is important to be aware of them. The find_single_gene_imodulons function identifies iModulons that are likely dominated by a single gene.\n",
    "\n",
    "The iModulons identified by ``find_single_gene_imodulons`` may contain more than one gene, since a threshold-agnostic method is used to identify these iModulons."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sg_imods = ica_data.find_single_gene_imodulons(save=True)\n",
    "len(sg_imods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i,mod in enumerate(sg_imods):\n",
    "    ica_data.rename_imodulons({mod:'SG_'+str(i+1)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'single_gene'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/90/3j1z360d3ys7y_4lx77mk_br0000gp/T/ipykernel_3875/1381282101.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mica_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimodulon_table\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mica_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mimodulon_table\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msingle_gene\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/imodulon/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5485\u001b[0m         ):\n\u001b[1;32m   5486\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5487\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5488\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5489\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'single_gene'"
     ]
    }
   ],
   "source": [
    "ica_data.imodulon_table[ica_data.imodulon_table.single_gene == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_weight</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>eggNOG_OGs</th>\n",
       "      <th>Description</th>\n",
       "      <th>GOs</th>\n",
       "      <th>EC</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>KEGG_Module</th>\n",
       "      <th>KEGG_Reaction</th>\n",
       "      <th>...</th>\n",
       "      <th>BiGG_Reaction</th>\n",
       "      <th>PFAMs</th>\n",
       "      <th>accession</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>COG</th>\n",
       "      <th>operon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>XNR_RS00455</th>\n",
       "      <td>-0.027750</td>\n",
       "      <td>gloA</td>\n",
       "      <td>COG0346@1|root,COG0346@2|Bacteria,2IHZH@201174...</td>\n",
       "      <td>glyoxalase bleomycin resistance protein dioxyg...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>Glyoxalase</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_0096</td>\n",
       "      <td>113835</td>\n",
       "      <td>114227</td>\n",
       "      <td>+</td>\n",
       "      <td>VOC family protein</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>Op78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS00460</th>\n",
       "      <td>-0.026695</td>\n",
       "      <td>XNR_RS00460</td>\n",
       "      <td>COG1846@1|root,COG1846@2|Bacteria,2GM7K@201174...</td>\n",
       "      <td>MarR family</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>MarR_2</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_0097</td>\n",
       "      <td>114261</td>\n",
       "      <td>114791</td>\n",
       "      <td>-</td>\n",
       "      <td>MarR family transcriptional regulator</td>\n",
       "      <td>Transcription</td>\n",
       "      <td>Op79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS00470</th>\n",
       "      <td>-0.035224</td>\n",
       "      <td>XNR_RS00470</td>\n",
       "      <td>290IR@1|root,2ZN7H@2|Bacteria,2IFZ1@201174|Act...</td>\n",
       "      <td>Domain of unknown function (DUF5134)</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>DUF5134</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_0099</td>\n",
       "      <td>116714</td>\n",
       "      <td>117280</td>\n",
       "      <td>+</td>\n",
       "      <td>DUF5134 domain-containing protein</td>\n",
       "      <td>Function unknown</td>\n",
       "      <td>Op80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS00540</th>\n",
       "      <td>0.045767</td>\n",
       "      <td>XNR_RS00540</td>\n",
       "      <td>COG2971@1|root,COG2971@2|Bacteria,2GKBF@201174...</td>\n",
       "      <td>BadF BadG BcrA BcrD</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>BcrAD_BadFG</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_0113</td>\n",
       "      <td>130210</td>\n",
       "      <td>131250</td>\n",
       "      <td>+</td>\n",
       "      <td>ATPase BadF/BadG/BcrA/BcrD type</td>\n",
       "      <td>Carbohydrate transport and metabolism</td>\n",
       "      <td>Op94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS00625</th>\n",
       "      <td>-0.038861</td>\n",
       "      <td>XNR_RS00625</td>\n",
       "      <td>COG4291@1|root,COG4291@2|Bacteria,2IIVD@201174...</td>\n",
       "      <td>membrane</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>DUF1345</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_0130</td>\n",
       "      <td>154417</td>\n",
       "      <td>155052</td>\n",
       "      <td>+</td>\n",
       "      <td>DUF1345 domain-containing protein</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>Op108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS28205</th>\n",
       "      <td>0.033241</td>\n",
       "      <td>XNR_RS28205</td>\n",
       "      <td>COG1280@1|root,COG1280@2|Bacteria,2GMDP@201174...</td>\n",
       "      <td>Lysine exporter protein (LysE YggA)</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>LysE</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_5687</td>\n",
       "      <td>6414812</td>\n",
       "      <td>6415465</td>\n",
       "      <td>+</td>\n",
       "      <td>LysE family translocator</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>Op4258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS31650</th>\n",
       "      <td>-0.058175</td>\n",
       "      <td>XNR_RS31650</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_5714</td>\n",
       "      <td>6438487</td>\n",
       "      <td>6438621</td>\n",
       "      <td>+</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>X</td>\n",
       "      <td>Op4277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS31655</th>\n",
       "      <td>-0.034436</td>\n",
       "      <td>XNR_RS31655</td>\n",
       "      <td>COG0654@1|root,COG0654@2|Bacteria,2GN2W@201174...</td>\n",
       "      <td>PFAM monooxygenase FAD-binding</td>\n",
       "      <td>-</td>\n",
       "      <td>1.14.13.232,1.14.13.233,1.14.13.7</td>\n",
       "      <td>ko:K03380,ko:K14252</td>\n",
       "      <td>ko00253,ko00623,ko00627,ko01057,ko01120,ko0113...</td>\n",
       "      <td>M00780,M00823</td>\n",
       "      <td>R00815,R03566,R05462,R09190</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>FAD_binding_3,Phe_hydrox_dim</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6497081</td>\n",
       "      <td>6497611</td>\n",
       "      <td>-</td>\n",
       "      <td>FAD-dependent monooxygenase</td>\n",
       "      <td>Energy production and conversion</td>\n",
       "      <td>Op4309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS28855</th>\n",
       "      <td>0.032170</td>\n",
       "      <td>dppC</td>\n",
       "      <td>COG1173@1|root,COG1173@2|Bacteria,2GKAW@201174...</td>\n",
       "      <td>PFAM binding-protein-dependent transport syste...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K02031,ko:K02034</td>\n",
       "      <td>ko02024,map02024</td>\n",
       "      <td>M00239</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>ABC_tran,BPD_transp_1</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_5822</td>\n",
       "      <td>6555690</td>\n",
       "      <td>6556559</td>\n",
       "      <td>+</td>\n",
       "      <td>ABC transporter permease</td>\n",
       "      <td>Amino acid transport and metabolism</td>\n",
       "      <td>Op4344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XNR_RS29305</th>\n",
       "      <td>-0.030863</td>\n",
       "      <td>XNR_RS29305</td>\n",
       "      <td>2E5BW@1|root,3303Y@2|Bacteria,2IKUN@201174|Act...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_020990.1</td>\n",
       "      <td>XNR_5912</td>\n",
       "      <td>6789753</td>\n",
       "      <td>6790067</td>\n",
       "      <td>+</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>-</td>\n",
       "      <td>Op4408</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>248 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_weight    gene_name  \\\n",
       "XNR_RS00455    -0.027750         gloA   \n",
       "XNR_RS00460    -0.026695  XNR_RS00460   \n",
       "XNR_RS00470    -0.035224  XNR_RS00470   \n",
       "XNR_RS00540     0.045767  XNR_RS00540   \n",
       "XNR_RS00625    -0.038861  XNR_RS00625   \n",
       "...                  ...          ...   \n",
       "XNR_RS28205     0.033241  XNR_RS28205   \n",
       "XNR_RS31650    -0.058175  XNR_RS31650   \n",
       "XNR_RS31655    -0.034436  XNR_RS31655   \n",
       "XNR_RS28855     0.032170         dppC   \n",
       "XNR_RS29305    -0.030863  XNR_RS29305   \n",
       "\n",
       "                                                    eggNOG_OGs  \\\n",
       "XNR_RS00455  COG0346@1|root,COG0346@2|Bacteria,2IHZH@201174...   \n",
       "XNR_RS00460  COG1846@1|root,COG1846@2|Bacteria,2GM7K@201174...   \n",
       "XNR_RS00470  290IR@1|root,2ZN7H@2|Bacteria,2IFZ1@201174|Act...   \n",
       "XNR_RS00540  COG2971@1|root,COG2971@2|Bacteria,2GKBF@201174...   \n",
       "XNR_RS00625  COG4291@1|root,COG4291@2|Bacteria,2IIVD@201174...   \n",
       "...                                                        ...   \n",
       "XNR_RS28205  COG1280@1|root,COG1280@2|Bacteria,2GMDP@201174...   \n",
       "XNR_RS31650                                                NaN   \n",
       "XNR_RS31655  COG0654@1|root,COG0654@2|Bacteria,2GN2W@201174...   \n",
       "XNR_RS28855  COG1173@1|root,COG1173@2|Bacteria,2GKAW@201174...   \n",
       "XNR_RS29305  2E5BW@1|root,3303Y@2|Bacteria,2IKUN@201174|Act...   \n",
       "\n",
       "                                                   Description  GOs  \\\n",
       "XNR_RS00455  glyoxalase bleomycin resistance protein dioxyg...    -   \n",
       "XNR_RS00460                                        MarR family    -   \n",
       "XNR_RS00470               Domain of unknown function (DUF5134)    -   \n",
       "XNR_RS00540                                BadF BadG BcrA BcrD    -   \n",
       "XNR_RS00625                                           membrane    -   \n",
       "...                                                        ...  ...   \n",
       "XNR_RS28205                Lysine exporter protein (LysE YggA)    -   \n",
       "XNR_RS31650                                                NaN  NaN   \n",
       "XNR_RS31655                     PFAM monooxygenase FAD-binding    -   \n",
       "XNR_RS28855  PFAM binding-protein-dependent transport syste...    -   \n",
       "XNR_RS29305                                                  -    -   \n",
       "\n",
       "                                            EC              KEGG_ko  \\\n",
       "XNR_RS00455                                  -                    -   \n",
       "XNR_RS00460                                  -                    -   \n",
       "XNR_RS00470                                  -                    -   \n",
       "XNR_RS00540                                  -                    -   \n",
       "XNR_RS00625                                  -                    -   \n",
       "...                                        ...                  ...   \n",
       "XNR_RS28205                                  -                    -   \n",
       "XNR_RS31650                                NaN                  NaN   \n",
       "XNR_RS31655  1.14.13.232,1.14.13.233,1.14.13.7  ko:K03380,ko:K14252   \n",
       "XNR_RS28855                                  -  ko:K02031,ko:K02034   \n",
       "XNR_RS29305                                  -                    -   \n",
       "\n",
       "                                                  KEGG_Pathway    KEGG_Module  \\\n",
       "XNR_RS00455                                                  -              -   \n",
       "XNR_RS00460                                                  -              -   \n",
       "XNR_RS00470                                                  -              -   \n",
       "XNR_RS00540                                                  -              -   \n",
       "XNR_RS00625                                                  -              -   \n",
       "...                                                        ...            ...   \n",
       "XNR_RS28205                                                  -              -   \n",
       "XNR_RS31650                                                NaN            NaN   \n",
       "XNR_RS31655  ko00253,ko00623,ko00627,ko01057,ko01120,ko0113...  M00780,M00823   \n",
       "XNR_RS28855                                   ko02024,map02024         M00239   \n",
       "XNR_RS29305                                                  -              -   \n",
       "\n",
       "                           KEGG_Reaction  ... BiGG_Reaction  \\\n",
       "XNR_RS00455                            -  ...             -   \n",
       "XNR_RS00460                            -  ...             -   \n",
       "XNR_RS00470                            -  ...             -   \n",
       "XNR_RS00540                            -  ...             -   \n",
       "XNR_RS00625                            -  ...             -   \n",
       "...                                  ...  ...           ...   \n",
       "XNR_RS28205                            -  ...             -   \n",
       "XNR_RS31650                          NaN  ...           NaN   \n",
       "XNR_RS31655  R00815,R03566,R05462,R09190  ...             -   \n",
       "XNR_RS28855                            -  ...             -   \n",
       "XNR_RS29305                            -  ...             -   \n",
       "\n",
       "                                    PFAMs    accession old_locus_tag    start  \\\n",
       "XNR_RS00455                    Glyoxalase  NC_020990.1      XNR_0096   113835   \n",
       "XNR_RS00460                        MarR_2  NC_020990.1      XNR_0097   114261   \n",
       "XNR_RS00470                       DUF5134  NC_020990.1      XNR_0099   116714   \n",
       "XNR_RS00540                   BcrAD_BadFG  NC_020990.1      XNR_0113   130210   \n",
       "XNR_RS00625                       DUF1345  NC_020990.1      XNR_0130   154417   \n",
       "...                                   ...          ...           ...      ...   \n",
       "XNR_RS28205                          LysE  NC_020990.1      XNR_5687  6414812   \n",
       "XNR_RS31650                           NaN  NC_020990.1      XNR_5714  6438487   \n",
       "XNR_RS31655  FAD_binding_3,Phe_hydrox_dim  NC_020990.1           NaN  6497081   \n",
       "XNR_RS28855         ABC_tran,BPD_transp_1  NC_020990.1      XNR_5822  6555690   \n",
       "XNR_RS29305                             -  NC_020990.1      XNR_5912  6789753   \n",
       "\n",
       "                 end strand                           gene_product  \\\n",
       "XNR_RS00455   114227      +                     VOC family protein   \n",
       "XNR_RS00460   114791      -  MarR family transcriptional regulator   \n",
       "XNR_RS00470   117280      +      DUF5134 domain-containing protein   \n",
       "XNR_RS00540   131250      +        ATPase BadF/BadG/BcrA/BcrD type   \n",
       "XNR_RS00625   155052      +      DUF1345 domain-containing protein   \n",
       "...              ...    ...                                    ...   \n",
       "XNR_RS28205  6415465      +               LysE family translocator   \n",
       "XNR_RS31650  6438621      +                   hypothetical protein   \n",
       "XNR_RS31655  6497611      -            FAD-dependent monooxygenase   \n",
       "XNR_RS28855  6556559      +               ABC transporter permease   \n",
       "XNR_RS29305  6790067      +                   hypothetical protein   \n",
       "\n",
       "                                               COG  operon  \n",
       "XNR_RS00455    Amino acid transport and metabolism    Op78  \n",
       "XNR_RS00460                          Transcription    Op79  \n",
       "XNR_RS00470                       Function unknown    Op80  \n",
       "XNR_RS00540  Carbohydrate transport and metabolism    Op94  \n",
       "XNR_RS00625       Energy production and conversion   Op108  \n",
       "...                                            ...     ...  \n",
       "XNR_RS28205    Amino acid transport and metabolism  Op4258  \n",
       "XNR_RS31650                                      X  Op4277  \n",
       "XNR_RS31655       Energy production and conversion  Op4309  \n",
       "XNR_RS28855    Amino acid transport and metabolism  Op4344  \n",
       "XNR_RS29305                                      -  Op4408  \n",
       "\n",
       "[248 rows x 24 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.view_imodulon(17)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save iModulon object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.util import explained_variance\n",
    "from pymodulon.io import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add iModulon sizes and explained variance\n",
    "for im in ica_data.imodulon_names:\n",
    "    ica_data.imodulon_table.loc[im,'imodulon_size'] = len(ica_data.view_imodulon(im))\n",
    "    ica_data.imodulon_table.loc[im,'explained_variance'] = explained_variance(ica_data,imodulons=im)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>explained_variance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>97.0</td>\n",
       "      <td>0.006052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>24.0</td>\n",
       "      <td>0.003215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16.0</td>\n",
       "      <td>0.004825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>21.0</td>\n",
       "      <td>0.010579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>17.0</td>\n",
       "      <td>0.002131</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   imodulon_size  explained_variance\n",
       "0           97.0            0.006052\n",
       "1           24.0            0.003215\n",
       "2           16.0            0.004825\n",
       "3           21.0            0.010579\n",
       "4           17.0            0.002131"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ica_data.imodulon_table.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This will save your iModulon table, your thresholds, and any other information stored in the ica_data object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_to_json(ica_data, path.join('../..','data','interim','salb_raw.json.gz'))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you prefer to view and edit your iModulon table in excel, save it as a CSV and reload the iModulon as before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "ica_data.imodulon_table.to_csv(path.join('../..','data','interim','imodulon_table_raw.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>imodulon_size</th>\n",
       "      <th>explained_variance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>0.006052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.003215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.004825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.010579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.002131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>58</td>\n",
       "      <td>514.0</td>\n",
       "      <td>0.036333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>59</td>\n",
       "      <td>42.0</td>\n",
       "      <td>0.004464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>60</td>\n",
       "      <td>411.0</td>\n",
       "      <td>0.002833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>61</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.000503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>62</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.000852</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0  imodulon_size  explained_variance\n",
       "0            0           97.0            0.006052\n",
       "1            1           24.0            0.003215\n",
       "2            2           16.0            0.004825\n",
       "3            3           21.0            0.010579\n",
       "4            4           17.0            0.002131\n",
       "..         ...            ...                 ...\n",
       "58          58          514.0            0.036333\n",
       "59          59           42.0            0.004464\n",
       "60          60          411.0            0.002833\n",
       "61          61            6.0            0.000503\n",
       "62          62            2.0            0.000852\n",
       "\n",
       "[63 rows x 3 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imodulon_names_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the CSV file containing the imodulon names\n",
    "imodulon_names_df = pd.read_csv(path.join(interim_data, 'imodulon_table_raw.csv'))\n",
    "# Extract the names into a list\n",
    "imodulon_names = imodulon_names_df['Unnamed: 0'].tolist()\n",
    "all_genes = []  # List to store all genes from each imodulon\n",
    "\n",
    "for imodulon_name in imodulon_names:\n",
    "    imodulome_data = ica_data.view_imodulon(imodulon_name)\n",
    "    genes = imodulome_data['gene_name'].tolist()\n",
    "    for gene in genes:\n",
    "        all_genes.append((gene, imodulon_name))\n",
    "\n",
    "# Convert the collected gene data to a pandas DataFrame\n",
    "df = pd.DataFrame(all_genes, columns=['Gene', 'Imodulon'])\n",
    "\n",
    "# Save the DataFrame to a CSV file\n",
    "df.to_csv(path.join(data_dir,'genes_with_imodulon.csv'), index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
